// @microsoft/teamsfx-core/build/component/generator/templates/ssrTabGenerator.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsrTabGenerator = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const templateGenerator_1 = require("./templateGenerator");
const question_1 = require("../../../question");
const templateNames_1 = require("./templateNames");
// For the APS.NET server-side rendering tab
class SsrTabGenerator extends templateGenerator_1.DefaultTemplateGenerator {
    constructor() {
        super(...arguments);
        this.capabilities2TemplateNames = {
            [question_1.CapabilityOptions.nonSsoTab().id]: templateNames_1.TemplateNames.TabSSR,
            [question_1.CapabilityOptions.tab().id]: templateNames_1.TemplateNames.SsoTabSSR,
        };
    }
    activate(context, inputs) {
        const capability = inputs.capabilities;
        return (this.capabilities2TemplateNames[capability] !== undefined &&
            inputs[question_1.QuestionNames.ProgrammingLanguage] === question_1.ProgrammingLanguage.CSharp &&
            inputs.targetFramework !== "net6.0" &&
            inputs.targetFramework !== "net7.0");
    }
    getTemplateInfos(context, inputs, destinationPath) {
        return Promise.resolve(teamsfx_api_1.ok([
            {
                templateName: this.capabilities2TemplateNames[inputs.capabilities],
                language: question_1.ProgrammingLanguage.CSharp,
            },
        ]));
    }
}
exports.SsrTabGenerator = SsrTabGenerator;
