// @microsoft/teamsfx-core/build/component/generator/templates/templateNames.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.inputsToTemplateName = exports.getTemplateName = exports.tryGetTemplateName = exports.Feature2TemplateName = exports.TemplateNames = void 0;
const create_1 = require("../../../question/create");
const questionNames_1 = require("../../../question/questionNames");
var TemplateNames;
(function (TemplateNames) {
    TemplateNames["Tab"] = "non-sso-tab";
    TemplateNames["SsoTab"] = "sso-tab";
    TemplateNames["SsoTabObo"] = "sso-tab-with-obo-flow";
    TemplateNames["TabSSR"] = "non-sso-tab-ssr";
    TemplateNames["SsoTabSSR"] = "sso-tab-ssr";
    TemplateNames["DashboardTab"] = "dashboard-tab";
    TemplateNames["NotificationRestify"] = "notification-restify";
    TemplateNames["NotificationWebApi"] = "notification-webapi";
    TemplateNames["NotificationHttpTriggerIsolated"] = "notification-http-trigger-isolated";
    TemplateNames["NotificationHttpTrigger"] = "notification-http-trigger";
    TemplateNames["NotificationTimerTriggerIsolated"] = "notification-timer-trigger-isolated";
    TemplateNames["NotificationTimerTrigger"] = "notification-timer-trigger";
    TemplateNames["NotificationHttpTimerTriggerIsolated"] = "notification-http-timer-trigger-isolated";
    TemplateNames["NotificationHttpTimerTrigger"] = "notification-http-timer-trigger";
    TemplateNames["CommandAndResponse"] = "command-and-response";
    TemplateNames["Workflow"] = "workflow";
    TemplateNames["DefaultBot"] = "default-bot";
    TemplateNames["MessageExtension"] = "message-extension";
    TemplateNames["MessageExtensionAction"] = "message-extension-action";
    TemplateNames["MessageExtensionSearch"] = "message-extension-search";
    TemplateNames["MessageExtensionCopilot"] = "message-extension-copilot";
    TemplateNames["M365MessageExtension"] = "m365-message-extension";
    TemplateNames["TabAndDefaultBot"] = "non-sso-tab-default-bot";
    TemplateNames["BotAndMessageExtension"] = "default-bot-message-extension";
    TemplateNames["LinkUnfurling"] = "link-unfurling";
    TemplateNames["AIBot"] = "ai-bot";
    TemplateNames["AIAssistantBot"] = "ai-assistant-bot";
    TemplateNames["ApiPluginFromScratch"] = "api-plugin-from-scratch";
    TemplateNames["CopilotPluginFromScratch"] = "copilot-plugin-from-scratch";
    TemplateNames["CopilotPluginFromScratchApiKey"] = "copilot-plugin-from-scratch-api-key";
    TemplateNames["ApiMessageExtensionSso"] = "api-message-extension-sso";
    TemplateNames["CustomCopilotBasic"] = "custom-copilot-basic";
    TemplateNames["CustomCopilotRagCustomize"] = "custom-copilot-rag-customize";
    TemplateNames["CustomCopilotRagAzureAISearch"] = "custom-copilot-rag-azure-ai-search";
    TemplateNames["CustomCopilotRagCustomApi"] = "custom-copilot-rag-custom-api";
    TemplateNames["CustomCopilotRagMicrosoft365"] = "custom-copilot-rag-microsoft365";
    TemplateNames["CustomCopilotAssistantNew"] = "custom-copilot-assistant-new";
    TemplateNames["CustomCopilotAssistantAssistantsApi"] = "custom-copilot-assistant-assistants-api";
    TemplateNames["BasicGpt"] = "copilot-gpt-basic";
    TemplateNames["GptWithPluginFromScratch"] = "copilot-gpt-from-scratch-plugin";
})(TemplateNames = exports.TemplateNames || (exports.TemplateNames = {}));
// TODO: remove this mapping after all generators are migrated to new generator pattern
exports.Feature2TemplateName = {
    [`${create_1.CapabilityOptions.nonSsoTab().id}:undefined`]: TemplateNames.Tab,
    [`${create_1.CapabilityOptions.tab().id}:undefined`]: TemplateNames.SsoTab,
    [`${create_1.CapabilityOptions.m365SsoLaunchPage().id}:undefined`]: TemplateNames.SsoTabObo,
    [`${create_1.CapabilityOptions.nonSsoTab().id}:ssr`]: TemplateNames.TabSSR,
    [`${create_1.CapabilityOptions.tab().id}:ssr`]: TemplateNames.SsoTabSSR,
    [`${create_1.CapabilityOptions.dashboardTab().id}:undefined`]: TemplateNames.DashboardTab,
    [`${create_1.CapabilityOptions.notificationBot().id}:${create_1.NotificationTriggerOptions.appService().id}`]: TemplateNames.NotificationRestify,
    [`${create_1.CapabilityOptions.notificationBot().id}:${create_1.NotificationTriggerOptions.appServiceForVS().id}`]: TemplateNames.NotificationWebApi,
    [`${create_1.CapabilityOptions.notificationBot().id}:${create_1.NotificationTriggerOptions.functionsHttpTrigger().id}`]: TemplateNames.NotificationHttpTrigger,
    [`${create_1.CapabilityOptions.notificationBot().id}:${create_1.NotificationTriggerOptions.functionsHttpTriggerIsolated().id}`]: TemplateNames.NotificationHttpTriggerIsolated,
    [`${create_1.CapabilityOptions.notificationBot().id}:${create_1.NotificationTriggerOptions.functionsTimerTrigger().id}`]: TemplateNames.NotificationTimerTrigger,
    [`${create_1.CapabilityOptions.notificationBot().id}:${create_1.NotificationTriggerOptions.functionsTimerTriggerIsolated().id}`]: TemplateNames.NotificationTimerTriggerIsolated,
    [`${create_1.CapabilityOptions.notificationBot().id}:${create_1.NotificationTriggerOptions.functionsHttpAndTimerTrigger().id}`]: TemplateNames.NotificationHttpTimerTrigger,
    [`${create_1.CapabilityOptions.notificationBot().id}:${create_1.NotificationTriggerOptions.functionsHttpAndTimerTriggerIsolated().id}`]: TemplateNames.NotificationHttpTimerTriggerIsolated,
    [`${create_1.CapabilityOptions.commandBot().id}:undefined`]: TemplateNames.CommandAndResponse,
    [`${create_1.CapabilityOptions.workflowBot().id}:undefined`]: TemplateNames.Workflow,
    [`${create_1.CapabilityOptions.basicBot().id}:undefined`]: TemplateNames.DefaultBot,
    [`${create_1.CapabilityOptions.me().id}:undefined`]: TemplateNames.MessageExtension,
    [`${create_1.CapabilityOptions.collectFormMe().id}:undefined`]: TemplateNames.MessageExtensionAction,
    [`${create_1.CapabilityOptions.SearchMe().id}:undefined`]: TemplateNames.MessageExtensionSearch,
    [`${create_1.CapabilityOptions.m365SearchMe().id}:undefined:${create_1.MeArchitectureOptions.botPlugin().id}`]: TemplateNames.MessageExtensionCopilot,
    [`${create_1.CapabilityOptions.m365SearchMe().id}:undefined:${create_1.MeArchitectureOptions.botMe().id}`]: TemplateNames.M365MessageExtension,
    [`${create_1.CapabilityOptions.nonSsoTabAndBot().id}:undefined`]: TemplateNames.TabAndDefaultBot,
    [`${create_1.CapabilityOptions.botAndMe().id}:undefined`]: TemplateNames.BotAndMessageExtension,
    [`${create_1.CapabilityOptions.linkUnfurling().id}:undefined`]: TemplateNames.LinkUnfurling,
    [`${create_1.CapabilityOptions.aiBot().id}:undefined`]: TemplateNames.AIBot,
    [`${create_1.CapabilityOptions.aiAssistantBot().id}:undefined`]: TemplateNames.AIAssistantBot,
    [`${create_1.CapabilityOptions.copilotPluginNewApi().id}:undefined`]: TemplateNames.ApiPluginFromScratch,
    [`${create_1.CapabilityOptions.m365SearchMe().id}:undefined:${create_1.MeArchitectureOptions.newApi().id}:${create_1.ApiMessageExtensionAuthOptions.none().id}`]: TemplateNames.CopilotPluginFromScratch,
    [`${create_1.CapabilityOptions.m365SearchMe().id}:undefined:${create_1.MeArchitectureOptions.newApi().id}:${create_1.ApiMessageExtensionAuthOptions.apiKey().id}`]: TemplateNames.CopilotPluginFromScratchApiKey,
    [`${create_1.CapabilityOptions.m365SearchMe().id}:undefined:${create_1.MeArchitectureOptions.newApi().id}:${create_1.ApiMessageExtensionAuthOptions.microsoftEntra().id}`]: TemplateNames.ApiMessageExtensionSso,
    [`${create_1.CapabilityOptions.customCopilotBasic().id}:undefined`]: TemplateNames.CustomCopilotBasic,
    [`${create_1.CapabilityOptions.customCopilotRag().id}:undefined:${create_1.CustomCopilotRagOptions.customize().id}`]: TemplateNames.CustomCopilotRagCustomize,
    [`${create_1.CapabilityOptions.customCopilotRag().id}:undefined:${create_1.CustomCopilotRagOptions.azureAISearch().id}`]: TemplateNames.CustomCopilotRagAzureAISearch,
    [`${create_1.CapabilityOptions.customCopilotRag().id}:undefined:${create_1.CustomCopilotRagOptions.customApi().id}`]: TemplateNames.CustomCopilotRagCustomApi,
    [`${create_1.CapabilityOptions.customCopilotRag().id}:undefined:${create_1.CustomCopilotRagOptions.microsoft365().id}`]: TemplateNames.CustomCopilotRagMicrosoft365,
    [`${create_1.CapabilityOptions.customCopilotAssistant().id}:undefined:${create_1.CustomCopilotAssistantOptions.new().id}`]: TemplateNames.CustomCopilotAssistantNew,
    [`${create_1.CapabilityOptions.customCopilotAssistant().id}:undefined:${create_1.CustomCopilotAssistantOptions.assistantsApi().id}`]: TemplateNames.CustomCopilotAssistantAssistantsApi,
    [`${create_1.CapabilityOptions.customizeGptBasic().id}:undefined`]: TemplateNames.BasicGpt,
    [`${create_1.CapabilityOptions.customizeGptWithPlugin().id}:undefined`]: TemplateNames.GptWithPluginFromScratch,
};
function tryGetTemplateName(inputs) {
    for (const [key, value] of exports.inputsToTemplateName) {
        if (Object.keys(key).every((k) => key[k] === inputs[k])) {
            return value;
        }
    }
}
exports.tryGetTemplateName = tryGetTemplateName;
function getTemplateName(inputs) {
    const templateName = tryGetTemplateName(inputs);
    if (!templateName) {
        throw new Error("Template name not found");
    }
    return templateName;
}
exports.getTemplateName = getTemplateName;
// When multiple template name matches, only the top one will be picked.
exports.inputsToTemplateName = new Map([
    [{ [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.nonSsoTab().id }, TemplateNames.Tab],
    [{ [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.tab().id }, TemplateNames.SsoTab],
    [
        { [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.m365SsoLaunchPage().id },
        TemplateNames.SsoTabObo,
    ],
    [
        { [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.dashboardTab().id },
        TemplateNames.DashboardTab,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.notificationBot().id,
            [questionNames_1.QuestionNames.BotTrigger]: create_1.NotificationTriggerOptions.appService().id,
        },
        TemplateNames.NotificationRestify,
    ],
    [
        {
            [questionNames_1.QuestionNames.ProgrammingLanguage]: create_1.ProgrammingLanguage.CSharp,
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.notificationBot().id,
            [questionNames_1.QuestionNames.BotTrigger]: create_1.NotificationTriggerOptions.appServiceForVS().id,
        },
        TemplateNames.NotificationWebApi,
    ],
    [
        {
            [questionNames_1.QuestionNames.ProgrammingLanguage]: create_1.ProgrammingLanguage.CSharp,
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.notificationBot().id,
            [questionNames_1.QuestionNames.BotTrigger]: create_1.NotificationTriggerOptions.functionsHttpTrigger().id,
            ["isIsolated"]: true,
        },
        TemplateNames.NotificationHttpTriggerIsolated,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.notificationBot().id,
            [questionNames_1.QuestionNames.BotTrigger]: create_1.NotificationTriggerOptions.functionsHttpTrigger().id,
        },
        TemplateNames.NotificationHttpTrigger,
    ],
    [
        {
            [questionNames_1.QuestionNames.ProgrammingLanguage]: create_1.ProgrammingLanguage.CSharp,
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.notificationBot().id,
            [questionNames_1.QuestionNames.BotTrigger]: create_1.NotificationTriggerOptions.functionsTimerTrigger().id,
            ["isIsolated"]: true,
        },
        TemplateNames.NotificationTimerTriggerIsolated,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.notificationBot().id,
            [questionNames_1.QuestionNames.BotTrigger]: create_1.NotificationTriggerOptions.functionsTimerTrigger().id,
        },
        TemplateNames.NotificationTimerTrigger,
    ],
    [
        {
            [questionNames_1.QuestionNames.ProgrammingLanguage]: create_1.ProgrammingLanguage.CSharp,
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.notificationBot().id,
            [questionNames_1.QuestionNames.BotTrigger]: create_1.NotificationTriggerOptions.functionsHttpAndTimerTrigger().id,
            ["isIsolated"]: true,
        },
        TemplateNames.NotificationHttpTimerTriggerIsolated,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.notificationBot().id,
            [questionNames_1.QuestionNames.BotTrigger]: create_1.NotificationTriggerOptions.functionsHttpAndTimerTrigger().id,
        },
        TemplateNames.NotificationHttpTimerTrigger,
    ],
    [
        { [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.commandBot().id },
        TemplateNames.CommandAndResponse,
    ],
    [{ [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.workflowBot().id }, TemplateNames.Workflow],
    [{ [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.basicBot().id }, TemplateNames.DefaultBot],
    [{ [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.me().id }, TemplateNames.MessageExtension],
    [
        { [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.collectFormMe().id },
        TemplateNames.MessageExtensionAction,
    ],
    [
        { [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.SearchMe().id },
        TemplateNames.MessageExtensionSearch,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.m365SearchMe().id,
            [questionNames_1.QuestionNames.MeArchitectureType]: create_1.MeArchitectureOptions.botPlugin().id,
        },
        TemplateNames.MessageExtensionCopilot,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.m365SearchMe().id,
            [questionNames_1.QuestionNames.MeArchitectureType]: create_1.MeArchitectureOptions.botMe().id,
        },
        TemplateNames.M365MessageExtension,
    ],
    [
        { [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.nonSsoTabAndBot().id },
        TemplateNames.TabAndDefaultBot,
    ],
    [
        { [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.botAndMe().id },
        TemplateNames.BotAndMessageExtension,
    ],
    [
        { [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.linkUnfurling().id },
        TemplateNames.LinkUnfurling,
    ],
    [{ [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.aiBot().id }, TemplateNames.AIBot],
    [
        { [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.aiAssistantBot().id },
        TemplateNames.AIAssistantBot,
    ],
    [
        { [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.copilotPluginNewApi().id },
        TemplateNames.ApiPluginFromScratch,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.m365SearchMe().id,
            [questionNames_1.QuestionNames.MeArchitectureType]: create_1.MeArchitectureOptions.newApi().id,
            [questionNames_1.QuestionNames.ApiMEAuth]: create_1.ApiMessageExtensionAuthOptions.none().id,
        },
        TemplateNames.CopilotPluginFromScratch,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.m365SearchMe().id,
            [questionNames_1.QuestionNames.MeArchitectureType]: create_1.MeArchitectureOptions.newApi().id,
            [questionNames_1.QuestionNames.ApiMEAuth]: create_1.ApiMessageExtensionAuthOptions.apiKey().id,
        },
        TemplateNames.CopilotPluginFromScratchApiKey,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.m365SearchMe().id,
            [questionNames_1.QuestionNames.MeArchitectureType]: create_1.MeArchitectureOptions.newApi().id,
            [questionNames_1.QuestionNames.ApiMEAuth]: create_1.ApiMessageExtensionAuthOptions.microsoftEntra().id,
        },
        TemplateNames.ApiMessageExtensionSso,
    ],
    [
        { [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.customCopilotBasic().id },
        TemplateNames.CustomCopilotBasic,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.customCopilotRag().id,
            [questionNames_1.QuestionNames.CustomCopilotRag]: create_1.CustomCopilotRagOptions.customize().id,
        },
        TemplateNames.CustomCopilotRagCustomize,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.customCopilotRag().id,
            [questionNames_1.QuestionNames.CustomCopilotRag]: create_1.CustomCopilotRagOptions.azureAISearch().id,
        },
        TemplateNames.CustomCopilotRagAzureAISearch,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.customCopilotRag().id,
            [questionNames_1.QuestionNames.CustomCopilotRag]: create_1.CustomCopilotRagOptions.customApi().id,
        },
        TemplateNames.CustomCopilotRagCustomApi,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.customCopilotRag().id,
            [questionNames_1.QuestionNames.CustomCopilotRag]: create_1.CustomCopilotRagOptions.microsoft365().id,
        },
        TemplateNames.CustomCopilotRagMicrosoft365,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.customCopilotAssistant().id,
            [questionNames_1.QuestionNames.CustomCopilotAssistant]: create_1.CustomCopilotAssistantOptions.new().id,
        },
        TemplateNames.CustomCopilotAssistantNew,
    ],
    [
        {
            [questionNames_1.QuestionNames.Capabilities]: create_1.CapabilityOptions.customCopilotAssistant().id,
            [questionNames_1.QuestionNames.CustomCopilotAssistant]: create_1.CustomCopilotAssistantOptions.assistantsApi().id,
        },
        TemplateNames.CustomCopilotAssistantAssistantsApi,
    ],
]);
