// @microsoft/teamsfx-core/build/question/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.capabilitiesHavePythonOption = exports.copilotPluginOptionIds = exports.copilotPluginNewApiOptionId = exports.copilotPluginExistingApiOptionIds = exports.copilotPluginOpenAIPluginOptionId = exports.copilotPluginApiSpecOptionId = void 0;
exports.copilotPluginApiSpecOptionId = "copilot-plugin-existing-api";
exports.copilotPluginOpenAIPluginOptionId = "copilot-plugin-openai-plugin";
exports.copilotPluginExistingApiOptionIds = [
    exports.copilotPluginApiSpecOptionId,
    exports.copilotPluginOpenAIPluginOptionId,
];
exports.copilotPluginNewApiOptionId = "copilot-plugin-new-api";
exports.copilotPluginOptionIds = [
    exports.copilotPluginNewApiOptionId,
    exports.copilotPluginApiSpecOptionId,
    exports.copilotPluginOpenAIPluginOptionId,
];
exports.capabilitiesHavePythonOption = [
    "custom-copilot-basic",
    "custom-copilot-rag-azureAISearch",
    "custom-copilot-rag-customize",
    "custom-copilot-agent-new",
];
