// @microsoft/teamsfx-core/build/question/create.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.createProjectCliHelpNode = exports.createSampleProjectQuestionNode = exports.createProjectQuestionNode = exports.capabilitySubTree = exports.CustomCopilotAssistantOptions = exports.CustomCopilotRagOptions = exports.apiOperationQuestion = exports.apiMessageExtensionAuthQuestion = exports.openAIPluginManifestLocationQuestion = exports.apiSpecLocationQuestion = exports.ApiMessageExtensionAuthOptions = exports.RuntimeOptions = exports.appNameQuestion = exports.getSolutionName = exports.AppNamePattern = exports.folderQuestion = exports.programmingLanguageQuestion = exports.ProgrammingLanguage = exports.getLanguageOptions = exports.getOfficeAddinFramework = exports.getAddinFrameworkOptions = exports.officeAddinFrameworkQuestion = exports.officeAddinHostingQuestion = exports.SPFxImportFolderQuestion = exports.SPFxVersionOptionIds = exports.SPFxWebpartNameQuestion = exports.SPFxFrameworkQuestion = exports.SPFxPackageSelectQuestion = exports.NotificationTriggerOptions = exports.meArchitectureQuestion = exports.MeArchitectureOptions = exports.capabilityQuestion = exports.CapabilityOptions = exports.OfficeAddinHostOptions = exports.projectTypeQuestion = exports.ProjectTypeOptions = exports.ScratchOptions = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const jsonschema = tslib_1.__importStar(require("jsonschema"));
const lodash_1 = require("lodash");
const os = tslib_1.__importStar(require("os"));
const path = tslib_1.__importStar(require("path"));
const constants_1 = require("../common/constants");
const correlator_1 = require("../common/correlator");
const featureFlags_1 = require("../common/featureFlags");
const localizeUtils_1 = require("../common/localizeUtils");
const samples_1 = require("../common/samples");
const utils_1 = require("../common/utils");
const developerPortalScaffoldUtils_1 = require("../component/developerPortalScaffoldUtils");
const utils_2 = require("../component/driver/teamsApp/utils/utils");
const helper_1 = require("../component/generator/copilotPlugin/helper");
const projectConfig_1 = require("../component/generator/officeXMLAddin/projectConfig");
const error_1 = require("../component/generator/spfx/error");
const constants_2 = require("../component/generator/spfx/utils/constants");
const utils_3 = require("../component/generator/spfx/utils/utils");
const utils_4 = require("../component/utils");
const error_2 = require("../error");
const constants_3 = require("./constants");
const questionNames_1 = require("./questionNames");
const util_1 = require("./util");
class ScratchOptions {
    static yes() {
        return {
            id: "yes",
            label: localizeUtils_1.getLocalizedString("core.ScratchOptionYes.label"),
            detail: localizeUtils_1.getLocalizedString("core.ScratchOptionYes.detail"),
        };
    }
    static no() {
        return {
            id: "no",
            label: localizeUtils_1.getLocalizedString("core.ScratchOptionNo.label"),
            detail: localizeUtils_1.getLocalizedString("core.ScratchOptionNo.detail"),
        };
    }
    static all() {
        return [ScratchOptions.yes(), ScratchOptions.no()];
    }
}
exports.ScratchOptions = ScratchOptions;
class ProjectTypeOptions {
    static getCreateGroupName() {
        return featureFlags_1.isChatParticipantEnabled()
            ? localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.createGroup.title")
            : undefined;
    }
    static tab(platform) {
        return {
            id: "tab-type",
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(browser) " : ""}${localizeUtils_1.getLocalizedString("core.TabOption.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.tab.detail"),
            groupName: ProjectTypeOptions.getCreateGroupName(),
        };
    }
    static bot(platform) {
        return {
            id: "bot-type",
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(hubot) " : ""}${localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.bot.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.bot.detail"),
            groupName: ProjectTypeOptions.getCreateGroupName(),
        };
    }
    static me(platform) {
        return {
            id: "me-type",
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(symbol-keyword) " : ""}${localizeUtils_1.getLocalizedString("core.MessageExtensionOption.label")}`,
            detail: featureFlags_1.isCopilotPluginEnabled()
                ? localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.messageExtension.copilotEnabled.detail")
                : localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.messageExtension.detail"),
            groupName: ProjectTypeOptions.getCreateGroupName(),
        };
    }
    static outlookAddin(platform) {
        return {
            id: "outlook-addin-type",
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(mail) " : ""}${localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.outlookAddin.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.outlookAddin.detail"),
            groupName: ProjectTypeOptions.getCreateGroupName(),
        };
    }
    static officeXMLAddin(platform) {
        return {
            id: "office-xml-addin-type",
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(teamsfx-m365) " : ""}${localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.mainEntry.title")}`,
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.mainEntry.detail"),
            groupName: ProjectTypeOptions.getCreateGroupName(),
        };
    }
    static officeAddin(platform) {
        return {
            id: "office-addin-type",
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(extensions) " : ""}${localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.officeAddin.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.officeAddin.detail"),
            groupName: ProjectTypeOptions.getCreateGroupName(),
        };
    }
    static officeAddinAllIds(platform) {
        return [
            ProjectTypeOptions.officeAddin(platform).id,
            ProjectTypeOptions.officeXMLAddin(platform).id,
            ProjectTypeOptions.outlookAddin(platform).id,
        ];
    }
    static copilotPlugin(platform) {
        return {
            id: "copilot-plugin-type",
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(teamsfx-copilot-plugin) " : ""}${localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.copilotPlugin.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.copilotPlugin.detail"),
            groupName: ProjectTypeOptions.getCreateGroupName(),
        };
    }
    static customCopilot(platform) {
        return {
            id: "custom-copilot-type",
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(teamsfx-custom-copilot) " : ""}${localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.customCopilot.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.customCopilot.detail"),
            groupName: ProjectTypeOptions.getCreateGroupName(),
        };
    }
    static startWithGithubCopilot() {
        return {
            id: "start-with-github-copilot",
            label: `$(comment-discussion) ${localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.copilotHelp.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.copilotHelp.detail"),
            groupName: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.copilotGroup.title"),
        };
    }
    static customizeGpt() {
        return {
            id: "customize-gpt-type",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.declarativeCopilot.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.declarativeCopilot.title"),
            groupName: ProjectTypeOptions.getCreateGroupName(),
        };
    }
}
exports.ProjectTypeOptions = ProjectTypeOptions;
function projectTypeQuestion() {
    const staticOptions = [
        ProjectTypeOptions.bot(teamsfx_api_1.Platform.CLI),
        ProjectTypeOptions.tab(teamsfx_api_1.Platform.CLI),
        ProjectTypeOptions.me(teamsfx_api_1.Platform.CLI),
        ProjectTypeOptions.officeXMLAddin(teamsfx_api_1.Platform.CLI),
        ProjectTypeOptions.officeAddin(teamsfx_api_1.Platform.CLI),
        ProjectTypeOptions.outlookAddin(teamsfx_api_1.Platform.CLI),
    ];
    return {
        name: questionNames_1.QuestionNames.ProjectType,
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.title"),
        type: "singleSelect",
        staticOptions: staticOptions,
        dynamicOptions: (inputs) => {
            var _a;
            const staticOptions = [];
            if (teamsfx_api_1.CLIPlatforms.includes(inputs.platform) &&
                featureFlags_1.featureFlagManager.getBooleanValue(featureFlags_1.FeatureFlags.CustomizeGpt)) {
                // Show in CLI only
                staticOptions.push(ProjectTypeOptions.customizeGpt());
            }
            if (featureFlags_1.isApiCopilotPluginEnabled()) {
                staticOptions.push(ProjectTypeOptions.copilotPlugin(inputs.platform));
            }
            staticOptions.push(ProjectTypeOptions.customCopilot(inputs.platform));
            staticOptions.push(ProjectTypeOptions.bot(inputs.platform), ProjectTypeOptions.tab(inputs.platform), ProjectTypeOptions.me(inputs.platform));
            if (developerPortalScaffoldUtils_1.isFromDevPortal(inputs)) {
                const projectType = (_a = developerPortalScaffoldUtils_1.getProjectTypeAndCapability(inputs.teamsAppFromTdp)) === null || _a === void 0 ? void 0 : _a.projectType;
                if (projectType) {
                    return [projectType];
                }
            }
            else {
                if (inputs.agent === "office") {
                    //only for @office agent, officeXMLAddin are supported
                    staticOptions.push(ProjectTypeOptions.officeXMLAddin(inputs.platform));
                }
                else {
                    if (featureFlags_1.isOfficeJSONAddinEnabled()) {
                        staticOptions.push(ProjectTypeOptions.officeAddin(inputs.platform));
                    }
                    else {
                        staticOptions.push(ProjectTypeOptions.outlookAddin(inputs.platform));
                    }
                }
            }
            if (inputs.platform === teamsfx_api_1.Platform.VSCode &&
                featureFlags_1.isChatParticipantEnabled() &&
                !inputs.teamsAppFromTdp) {
                staticOptions.push(ProjectTypeOptions.startWithGithubCopilot());
            }
            return staticOptions;
        },
        placeholder: localizeUtils_1.getLocalizedString("core.getCreateNewOrFromSampleQuestion.placeholder"),
        forgetLastValue: true,
        skipSingleOption: true,
    };
}
exports.projectTypeQuestion = projectTypeQuestion;
class OfficeAddinHostOptions {
    static all(platform) {
        return [
            OfficeAddinHostOptions.outlook(platform),
            OfficeAddinHostOptions.word(),
            OfficeAddinHostOptions.excel(),
            OfficeAddinHostOptions.powerpoint(),
        ];
    }
    static outlook(platform) {
        return {
            id: "outlook",
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(mail) " : ""}${localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.outlookAddin.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.outlookAddin.detail"),
            data: "Outlook",
        };
    }
    static word() {
        return {
            id: "word",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.word.title"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.word.detail"),
            data: "Word",
        };
    }
    static excel() {
        return {
            id: "excel",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.excel.title"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.excel.detail"),
            data: "Excel",
        };
    }
    static powerpoint() {
        return {
            id: "powerpoint",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.powerpoint.title"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.powerpoint.detail"),
            data: "PowerPoint",
        };
    }
}
exports.OfficeAddinHostOptions = OfficeAddinHostOptions;
class CapabilityOptions {
    // bot
    static basicBot() {
        return {
            id: "bot",
            label: `${localizeUtils_1.getLocalizedString("core.BotNewUIOption.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.BotNewUIOption.detail"),
        };
    }
    static notificationBot() {
        return {
            // For default option, id and cliName must be the same
            id: "notification",
            label: `${localizeUtils_1.getLocalizedString("core.NotificationOption.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.NotificationOption.detail"),
            data: "https://aka.ms/teamsfx-send-notification",
            buttons: [
                {
                    iconPath: "file-symlink-file",
                    tooltip: localizeUtils_1.getLocalizedString("core.option.github"),
                    command: "fx-extension.openTutorial",
                },
            ],
        };
    }
    static commandBot() {
        return {
            // id must match cli `yargsHelp`
            id: "command-bot",
            label: `${localizeUtils_1.getLocalizedString("core.CommandAndResponseOption.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.CommandAndResponseOption.detail"),
            data: "https://aka.ms/teamsfx-create-command",
            buttons: [
                {
                    iconPath: "file-symlink-file",
                    tooltip: localizeUtils_1.getLocalizedString("core.option.github"),
                    command: "fx-extension.openTutorial",
                },
            ],
        };
    }
    static workflowBot(inputs) {
        const item = {
            // id must match cli `yargsHelp`
            id: "workflow-bot",
            label: `${localizeUtils_1.getLocalizedString("core.WorkflowOption.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.WorkflowOption.detail"),
            data: "https://aka.ms/teamsfx-create-workflow",
            buttons: [
                {
                    iconPath: "file-symlink-file",
                    tooltip: localizeUtils_1.getLocalizedString("core.option.github"),
                    command: "fx-extension.openTutorial",
                },
            ],
        };
        if (inputs === null || inputs === void 0 ? void 0 : inputs.inProductDoc) {
            item.data = "cardActionResponse";
            item.buttons = [
                {
                    iconPath: "file-code",
                    tooltip: localizeUtils_1.getLocalizedString("core.option.inProduct"),
                    command: "fx-extension.openTutorial",
                },
            ];
        }
        return item;
    }
    //tab
    static nonSsoTab() {
        return {
            id: "tab-non-sso",
            label: `${localizeUtils_1.getLocalizedString("core.TabNonSso.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.TabNonSso.detail"),
            description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.worksInOutlookM365"),
        };
    }
    static tab() {
        return {
            id: "tab",
            label: localizeUtils_1.getLocalizedString("core.TabOption.label"),
            description: localizeUtils_1.getLocalizedString("core.TabOption.description"),
            detail: localizeUtils_1.getLocalizedString("core.TabOption.detail"),
        };
    }
    static m365SsoLaunchPage() {
        return {
            id: "sso-launch-page",
            label: `${localizeUtils_1.getLocalizedString("core.M365SsoLaunchPageOptionItem.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.M365SsoLaunchPageOptionItem.detail"),
            description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.worksInOutlookM365"),
        };
    }
    static dashboardTab() {
        return {
            id: "dashboard-tab",
            label: `${localizeUtils_1.getLocalizedString("core.DashboardOption.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.DashboardOption.detail"),
            description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.worksInOutlookM365"),
            data: "https://aka.ms/teamsfx-dashboard-app",
            buttons: [
                {
                    iconPath: "file-symlink-file",
                    tooltip: localizeUtils_1.getLocalizedString("core.option.github"),
                    command: "fx-extension.openTutorial",
                },
            ],
        };
    }
    static SPFxTab() {
        return {
            id: "tab-spfx",
            label: localizeUtils_1.getLocalizedString("core.TabSPFxOption.labelNew"),
            description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.worksInOutlookM365"),
            detail: localizeUtils_1.getLocalizedString("core.TabSPFxOption.detailNew"),
        };
    }
    //message extension
    static linkUnfurling() {
        return {
            id: "link-unfurling",
            label: `${localizeUtils_1.getLocalizedString("core.LinkUnfurlingOption.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.LinkUnfurlingOption.detail"),
            description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.worksInOutlook"),
        };
    }
    static m365SearchMe() {
        return {
            id: "search-app",
            label: `${localizeUtils_1.getLocalizedString("core.M365SearchAppOptionItem.label")}`,
            detail: featureFlags_1.isCopilotPluginEnabled()
                ? localizeUtils_1.getLocalizedString("core.M365SearchAppOptionItem.copilot.detail")
                : localizeUtils_1.getLocalizedString("core.M365SearchAppOptionItem.detail"),
        };
    }
    static SearchMe() {
        return {
            id: "search-message-extension",
            label: `${localizeUtils_1.getLocalizedString("core.M365SearchAppOptionItem.label")}`,
            detail: localizeUtils_1.getLocalizedString("core.SearchAppOptionItem.detail"),
        };
    }
    static collectFormMe() {
        return {
            id: "collect-form-message-extension",
            label: `${localizeUtils_1.getLocalizedString("core.MessageExtensionOption.labelNew")}`,
            detail: localizeUtils_1.getLocalizedString("core.MessageExtensionOption.detail"),
        };
    }
    static me() {
        return {
            id: "message-extension",
            label: localizeUtils_1.getLocalizedString("core.MessageExtensionOption.label"),
            description: localizeUtils_1.getLocalizedString("core.MessageExtensionOption.description"),
            detail: localizeUtils_1.getLocalizedString("core.MessageExtensionOption.detail"),
        };
    }
    static bots(inputs) {
        if ((inputs === null || inputs === void 0 ? void 0 : inputs.platform) === teamsfx_api_1.Platform.VS) {
            return [
                CapabilityOptions.basicBot(),
                CapabilityOptions.aiBot(),
                CapabilityOptions.aiAssistantBot(),
                CapabilityOptions.notificationBot(),
                CapabilityOptions.commandBot(),
                CapabilityOptions.workflowBot(inputs),
            ];
        }
        return [
            CapabilityOptions.basicBot(),
            CapabilityOptions.notificationBot(),
            CapabilityOptions.commandBot(),
            CapabilityOptions.workflowBot(inputs),
        ];
    }
    static tabs() {
        return [
            CapabilityOptions.nonSsoTab(),
            CapabilityOptions.m365SsoLaunchPage(),
            CapabilityOptions.dashboardTab(),
            CapabilityOptions.SPFxTab(),
        ];
    }
    static dotnetCaps(inputs) {
        const capabilities = [
            ...CapabilityOptions.copilotPlugins(),
            ...CapabilityOptions.bots(inputs),
            CapabilityOptions.nonSsoTab(),
            CapabilityOptions.tab(),
            ...CapabilityOptions.collectMECaps(),
        ];
        if (featureFlags_1.isTdpTemplateCliTestEnabled()) {
            capabilities.push(CapabilityOptions.me());
        }
        return capabilities;
    }
    /**
     * Collect all capabilities for message extension, including dotnet and nodejs.
     * @returns OptionItem[] capability list
     */
    static collectMECaps() {
        return [
            CapabilityOptions.m365SearchMe(),
            CapabilityOptions.collectFormMe(),
            CapabilityOptions.SearchMe(),
            CapabilityOptions.linkUnfurling(),
        ];
    }
    static mes(inputs) {
        return inputs !== undefined && getRuntime(inputs) === RuntimeOptions.DotNet().id
            ? [
                CapabilityOptions.SearchMe(),
                CapabilityOptions.collectFormMe(),
                CapabilityOptions.linkUnfurling(),
            ]
            : [
                CapabilityOptions.m365SearchMe(),
                CapabilityOptions.collectFormMe(),
                CapabilityOptions.linkUnfurling(),
            ];
    }
    static officeAddinStaticCapabilities(host) {
        const items = [];
        for (const h of Object.keys(projectConfig_1.OfficeAddinProjectConfig)) {
            if (host && h !== host)
                continue;
            const hostValue = projectConfig_1.OfficeAddinProjectConfig[h];
            for (const capability of Object.keys(hostValue)) {
                const capabilityValue = hostValue[capability];
                items.push({
                    id: capability,
                    label: localizeUtils_1.getLocalizedString(capabilityValue.title),
                    detail: localizeUtils_1.getLocalizedString(capabilityValue.detail),
                });
            }
        }
        return items;
    }
    static officeAddinDynamicCapabilities(projectType, host) {
        const items = [];
        const isOutlookAddin = projectType === ProjectTypeOptions.outlookAddin().id;
        const isOfficeAddin = projectType === ProjectTypeOptions.officeAddin().id;
        const isOfficeXMLAddinForOutlook = projectType === ProjectTypeOptions.officeXMLAddin().id &&
            host === OfficeAddinHostOptions.outlook().id;
        const pushToItems = (option) => {
            const capabilityValue = projectConfig_1.OfficeAddinProjectConfig.json[option];
            items.push({
                id: option,
                label: localizeUtils_1.getLocalizedString(capabilityValue.title),
                detail: localizeUtils_1.getLocalizedString(capabilityValue.detail),
            });
        };
        if (isOutlookAddin || isOfficeAddin || isOfficeXMLAddinForOutlook) {
            pushToItems("json-taskpane");
            if (isOutlookAddin || isOfficeXMLAddinForOutlook) {
                items.push(CapabilityOptions.outlookAddinImport());
            }
            else if (isOfficeAddin) {
                items.push(CapabilityOptions.officeContentAddin());
                items.push(CapabilityOptions.officeAddinImport());
            }
        }
        else {
            if (host) {
                const hostValue = projectConfig_1.OfficeAddinProjectConfig[host];
                for (const capability of Object.keys(hostValue)) {
                    const capabilityValue = hostValue[capability];
                    items.push({
                        id: capability,
                        label: localizeUtils_1.getLocalizedString(capabilityValue.title),
                        detail: localizeUtils_1.getLocalizedString(capabilityValue.detail),
                    });
                }
            }
        }
        return items;
    }
    static copilotPlugins() {
        return [
            CapabilityOptions.copilotPluginNewApi(),
            CapabilityOptions.copilotPluginApiSpec(),
            // CapabilityOptions.copilotPluginOpenAIPlugin(),
        ];
    }
    static customCopilots() {
        return [
            CapabilityOptions.customCopilotBasic(),
            CapabilityOptions.customCopilotRag(),
            CapabilityOptions.customCopilotAssistant(),
        ];
    }
    static tdpIntegrationCapabilities() {
        // templates that are used by TDP integration only
        return [
            CapabilityOptions.me(),
            CapabilityOptions.botAndMe(),
            CapabilityOptions.nonSsoTabAndBot(),
        ];
    }
    static customizeGptOptions() {
        return [CapabilityOptions.customizeGptBasic(), CapabilityOptions.customizeGptWithPlugin()];
    }
    /**
     * static capability list, which does not depend on any feature flags
     */
    static staticAll(inputs) {
        const capabilityOptions = [
            ...CapabilityOptions.bots(inputs),
            ...CapabilityOptions.tabs(),
            ...CapabilityOptions.collectMECaps(),
            ...CapabilityOptions.copilotPlugins(),
            ...CapabilityOptions.customCopilots(),
            ...CapabilityOptions.tdpIntegrationCapabilities(),
            ...CapabilityOptions.customizeGptOptions(),
        ];
        capabilityOptions.push(...CapabilityOptions.officeAddinStaticCapabilities());
        return capabilityOptions;
    }
    /**
     * dynamic capability list, which depends on feature flags
     */
    static all(inputs) {
        const capabilityOptions = [
            ...CapabilityOptions.bots(inputs),
            ...CapabilityOptions.tabs(),
            ...CapabilityOptions.collectMECaps(),
        ];
        if (featureFlags_1.isApiCopilotPluginEnabled()) {
            capabilityOptions.push(...CapabilityOptions.copilotPlugins());
        }
        if (featureFlags_1.featureFlagManager.getBooleanValue(featureFlags_1.FeatureFlags.CustomizeGpt)) {
            capabilityOptions.push(...CapabilityOptions.customizeGptOptions());
        }
        capabilityOptions.push(...CapabilityOptions.customCopilots());
        if (featureFlags_1.isTdpTemplateCliTestEnabled()) {
            // test templates that are used by TDP integration only
            capabilityOptions.push(...CapabilityOptions.tdpIntegrationCapabilities());
        }
        capabilityOptions.push(...CapabilityOptions.officeAddinDynamicCapabilities(inputs === null || inputs === void 0 ? void 0 : inputs.projectType, inputs === null || inputs === void 0 ? void 0 : inputs.host));
        return capabilityOptions;
    }
    static outlookAddinImport() {
        return {
            id: "outlook-addin-import",
            label: localizeUtils_1.getLocalizedString("core.importAddin.label"),
            detail: localizeUtils_1.getLocalizedString("core.importAddin.detail"),
        };
    }
    static officeAddinImport() {
        return {
            id: "office-addin-import",
            label: localizeUtils_1.getLocalizedString("core.importOfficeAddin.label"),
            detail: localizeUtils_1.getLocalizedString("core.importAddin.detail"),
            description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.previewOnWindow"),
        };
    }
    static officeContentAddin() {
        return {
            id: "office-content-addin",
            label: localizeUtils_1.getLocalizedString("core.officeContentAddin.label"),
            detail: localizeUtils_1.getLocalizedString("core.officeContentAddin.detail"),
        };
    }
    // static officeXMLAddinHostOptionItems(host: string): OptionItem[] {
    //   return getOfficeXMLAddinHostProjectOptions(host).map((x) => ({
    //     id: x.proj,
    //     label: getLocalizedString(x.title),
    //     detail: getLocalizedString(x.detail),
    //   }));
    // }
    // static jsonAddinTaskpane(): OptionItem {
    //   return {
    //     id: "json-taskpane",
    //     label: getLocalizedString("core.newTaskpaneAddin.label"),
    //     detail: getLocalizedString("core.newTaskpaneAddin.detail"),
    //     description: getLocalizedString(
    //       "core.createProjectQuestion.option.description.previewOnWindow"
    //     ),
    //   };
    // }
    // static officeAddinItems(): OptionItem[] {
    //   return officeAddinJsonData.getProjectTemplateNames().map((template) => ({
    //     id: template,
    //     label: getLocalizedString(officeAddinJsonData.getProjectDisplayName(template)),
    //     detail: getLocalizedString(officeAddinJsonData.getProjectDetails(template)),
    //   }));
    // }
    static nonSsoTabAndBot() {
        return {
            id: "TabNonSsoAndBot",
            label: "", // No need to set display name as this option won't be shown in UI
        };
    }
    static botAndMe() {
        return {
            id: "BotAndMessageExtension",
            label: "", // No need to set display name as this option won't be shown in UI
        };
    }
    // copilot plugin
    static copilotPluginNewApi() {
        return {
            id: constants_3.copilotPluginNewApiOptionId,
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.copilotPluginNewApiOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.copilotPluginNewApiOption.detail"),
        };
    }
    static copilotPluginApiSpec() {
        return {
            id: constants_3.copilotPluginApiSpecOptionId,
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.copilotPluginApiSpecOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.copilotPluginApiSpecOption.detail"),
        };
    }
    static copilotPluginOpenAIPlugin() {
        return {
            id: constants_3.copilotPluginOpenAIPluginOptionId,
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.copilotPluginAIPluginOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.copilotPluginAIPluginOption.detail"),
        };
    }
    static aiBot() {
        return {
            id: "ai-bot",
            label: localizeUtils_1.getLocalizedString("core.aiBotOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.aiBotOption.detail"),
        };
    }
    static aiAssistantBot() {
        return {
            id: "ai-assistant-bot",
            label: localizeUtils_1.getLocalizedString("core.aiAssistantBotOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.aiAssistantBotOption.detail"),
            description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.preview"),
        };
    }
    // custom copilot
    static customCopilotBasic() {
        return {
            id: "custom-copilot-basic",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotBasicOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotBasicOption.detail"),
        };
    }
    static customCopilotRag() {
        return {
            id: "custom-copilot-rag",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRagOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRagOption.detail"),
        };
    }
    static customCopilotAssistant() {
        return {
            id: "custom-copilot-agent",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotAssistantOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotAssistantOption.detail"),
        };
    }
    // customize GPT
    static customizeGptBasic() {
        return {
            id: "basic-declarative-copilot",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.declarativeCopilotBasic.title"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.declarativeCopilotBasic.detail"),
        };
    }
    static customizeGptWithPlugin() {
        return {
            id: "declarative-copilot-with-plugin-from-scratch",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.declarativeCopilotWithPlugin.title"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.declarativeCopilotWithPlugin.detail"),
        };
    }
}
exports.CapabilityOptions = CapabilityOptions;
function capabilityQuestion() {
    return {
        name: questionNames_1.QuestionNames.Capabilities,
        title: (inputs) => {
            const projectType = inputs[questionNames_1.QuestionNames.ProjectType];
            switch (projectType) {
                case ProjectTypeOptions.bot().id:
                    return localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.bot.title");
                case ProjectTypeOptions.tab().id:
                    return localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.tab.title");
                case ProjectTypeOptions.me().id:
                    return localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.messageExtension.title");
                case ProjectTypeOptions.outlookAddin().id:
                case ProjectTypeOptions.officeAddin().id:
                case ProjectTypeOptions.officeXMLAddin().id: {
                    switch (inputs[questionNames_1.QuestionNames.OfficeAddinHost]) {
                        case OfficeAddinHostOptions.outlook().id:
                            return localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.outlookAddin.title");
                        case OfficeAddinHostOptions.word().id:
                            return localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.word.create.title");
                        case OfficeAddinHostOptions.excel().id:
                            return localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.excel.create.title");
                        case OfficeAddinHostOptions.powerpoint().id:
                            return localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.powerpoint.create.title");
                    }
                    return localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.officeAddin.title");
                }
                case ProjectTypeOptions.copilotPlugin().id:
                    return localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.copilotPlugin.title");
                case ProjectTypeOptions.customCopilot().id:
                    return localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.customCopilot.title");
                case ProjectTypeOptions.customizeGpt().id:
                    return localizeUtils_1.getLocalizedString("core.createProjectQuestion.declarativeCopilotType.title");
                default:
                    return localizeUtils_1.getLocalizedString("core.createCapabilityQuestion.titleNew");
            }
        },
        cliDescription: "Specifies the Microsoft Teams App capability.",
        cliName: questionNames_1.CliQuestionName.Capability,
        cliShortName: "c",
        cliChoiceListCommand: "teamsapp list templates",
        type: "singleSelect",
        staticOptions: CapabilityOptions.staticAll(),
        dynamicOptions: (inputs) => {
            var _a;
            // from dev portal
            if (developerPortalScaffoldUtils_1.isFromDevPortal(inputs)) {
                const capability = (_a = developerPortalScaffoldUtils_1.getProjectTypeAndCapability(inputs.teamsAppFromTdp)) === null || _a === void 0 ? void 0 : _a.templateId;
                if (capability) {
                    return [capability];
                }
            }
            // dotnet capabilities
            if (getRuntime(inputs) === RuntimeOptions.DotNet().id) {
                return CapabilityOptions.dotnetCaps(inputs);
            }
            if (inputs.nonInteractive && inputs.platform === teamsfx_api_1.Platform.CLI) {
                //cli non-interactive mode the choice list is the same as staticOptions
                return CapabilityOptions.all(inputs);
            }
            // nodejs capabilities
            const projectType = inputs[questionNames_1.QuestionNames.ProjectType];
            if (projectType === ProjectTypeOptions.bot().id) {
                return CapabilityOptions.bots(inputs);
            }
            else if (projectType === ProjectTypeOptions.tab().id) {
                return CapabilityOptions.tabs();
            }
            else if (projectType === ProjectTypeOptions.me().id) {
                return CapabilityOptions.mes();
            }
            else if (ProjectTypeOptions.officeAddinAllIds().includes(projectType)) {
                return CapabilityOptions.officeAddinDynamicCapabilities(projectType, inputs[questionNames_1.QuestionNames.OfficeAddinHost]);
            }
            else if (projectType === ProjectTypeOptions.copilotPlugin().id) {
                return CapabilityOptions.copilotPlugins();
            }
            else if (projectType === ProjectTypeOptions.customCopilot().id) {
                return CapabilityOptions.customCopilots();
            }
            else if (projectType === ProjectTypeOptions.customizeGpt().id) {
                return CapabilityOptions.customizeGptOptions();
            }
            else {
                return CapabilityOptions.all(inputs);
            }
        },
        placeholder: (inputs) => {
            if (inputs[questionNames_1.QuestionNames.ProjectType] === ProjectTypeOptions.copilotPlugin().id) {
                return localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.copilotPlugin.placeholder");
            }
            else if (inputs[questionNames_1.QuestionNames.ProjectType] === ProjectTypeOptions.customCopilot().id) {
                return localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.customCopilot.placeholder");
            }
            return localizeUtils_1.getLocalizedString("core.createCapabilityQuestion.placeholder");
        },
        forgetLastValue: true,
        skipSingleOption: true,
    };
}
exports.capabilityQuestion = capabilityQuestion;
class MeArchitectureOptions {
    static botMe() {
        return {
            id: "bot",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.botMessageExtension.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.botMessageExtension.detail"),
            description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.worksInOutlook"),
        };
    }
    static botPlugin() {
        return {
            id: "bot-plugin",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.botMessageExtension.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.botMessageExtension.detail"),
            description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.worksInOutlookCopilot"),
        };
    }
    static newApi() {
        return {
            id: "new-api",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.copilotPluginNewApiOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.messageExtensionNewApiOption.detail"),
        };
    }
    static apiSpec() {
        return {
            id: "api-spec",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.copilotPluginApiSpecOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.messageExtensionApiSpecOption.detail"),
        };
    }
    static all() {
        return [
            MeArchitectureOptions.newApi(),
            MeArchitectureOptions.apiSpec(),
            featureFlags_1.isCopilotPluginEnabled() ? MeArchitectureOptions.botPlugin() : MeArchitectureOptions.botMe(),
        ];
    }
    static staticAll() {
        return [
            MeArchitectureOptions.newApi(),
            MeArchitectureOptions.apiSpec(),
            MeArchitectureOptions.botPlugin(),
            MeArchitectureOptions.botMe(),
        ];
    }
}
exports.MeArchitectureOptions = MeArchitectureOptions;
function meArchitectureQuestion() {
    return {
        name: questionNames_1.QuestionNames.MeArchitectureType,
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.meArchitecture.title"),
        cliDescription: "Architecture of Search Based Message Extension.",
        cliShortName: "m",
        type: "singleSelect",
        staticOptions: MeArchitectureOptions.staticAll(),
        dynamicOptions: (inputs) => {
            return MeArchitectureOptions.all();
        },
        default: MeArchitectureOptions.newApi().id,
        placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.copilotPlugin.placeholder"),
        forgetLastValue: true,
        skipSingleOption: true,
    };
}
exports.meArchitectureQuestion = meArchitectureQuestion;
var HostType;
(function (HostType) {
    HostType["AppService"] = "app-service";
    HostType["Functions"] = "azure-functions";
})(HostType || (HostType = {}));
const NotificationTriggers = {
    HTTP: "http",
    TIMER: "timer",
};
class NotificationTriggerOptions {
    static appService() {
        return {
            id: "http-restify",
            hostType: HostType.AppService,
            label: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-restify.label"),
            description: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-restify.description"),
            detail: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-restify.detail"),
        };
    }
    static appServiceForVS() {
        return {
            id: "http-webapi",
            hostType: HostType.AppService,
            label: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-webapi.label"),
            description: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-webapi.description"),
            detail: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-webapi.detail"),
        };
    }
    // NOTE: id must be the sample as cliName to prevent parsing error for CLI default value.
    static functionsTimerTrigger() {
        return {
            id: "timer-functions",
            hostType: HostType.Functions,
            triggers: [NotificationTriggers.TIMER],
            label: localizeUtils_1.getLocalizedString("plugins.bot.triggers.timer-functions.label"),
            description: localizeUtils_1.getLocalizedString("plugins.bot.triggers.timer-functions.description"),
            detail: localizeUtils_1.getLocalizedString("plugins.bot.triggers.timer-functions.detail"),
        };
    }
    static functionsTimerTriggerIsolated() {
        return {
            id: "timer-functions-isolated",
            hostType: HostType.Functions,
            triggers: [NotificationTriggers.TIMER],
            label: localizeUtils_1.getLocalizedString("plugins.bot.triggers.timer-functions.label"),
            description: localizeUtils_1.getLocalizedString("plugins.bot.triggers.timer-functions.description"),
            detail: localizeUtils_1.getLocalizedString("plugins.bot.triggers.timer-functions.detail"),
        };
    }
    static functionsHttpAndTimerTrigger() {
        return {
            id: "http-and-timer-functions",
            hostType: HostType.Functions,
            triggers: [NotificationTriggers.HTTP, NotificationTriggers.TIMER],
            label: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-and-timer-functions.label"),
            description: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-and-timer-functions.description"),
            detail: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-and-timer-functions.detail"),
        };
    }
    static functionsHttpAndTimerTriggerIsolated() {
        return {
            id: "http-and-timer-functions-isolated",
            hostType: HostType.Functions,
            triggers: [NotificationTriggers.HTTP, NotificationTriggers.TIMER],
            label: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-and-timer-functions.label"),
            description: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-and-timer-functions.description"),
            detail: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-and-timer-functions.detail"),
        };
    }
    static functionsHttpTrigger() {
        return {
            id: "http-functions",
            hostType: HostType.Functions,
            triggers: [NotificationTriggers.HTTP],
            label: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-functions.label"),
            description: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-functions.description"),
            detail: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-functions.detail"),
        };
    }
    static functionsHttpTriggerIsolated() {
        return {
            id: "http-functions-isolated",
            hostType: HostType.Functions,
            triggers: [NotificationTriggers.HTTP],
            label: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-functions.label"),
            description: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-functions.description"),
            detail: localizeUtils_1.getLocalizedString("plugins.bot.triggers.http-functions.detail"),
        };
    }
    static functionsTriggers() {
        return [
            NotificationTriggerOptions.functionsHttpAndTimerTrigger(),
            NotificationTriggerOptions.functionsHttpTrigger(),
            NotificationTriggerOptions.functionsTimerTrigger(),
        ];
    }
    static all() {
        return [
            NotificationTriggerOptions.appService(),
            NotificationTriggerOptions.appServiceForVS(),
            NotificationTriggerOptions.functionsHttpAndTimerTrigger(),
            NotificationTriggerOptions.functionsHttpTrigger(),
            NotificationTriggerOptions.functionsTimerTrigger(),
        ];
    }
}
exports.NotificationTriggerOptions = NotificationTriggerOptions;
function getRuntime(inputs) {
    let runtime = RuntimeOptions.NodeJS().id;
    if (featureFlags_1.isCLIDotNetEnabled()) {
        runtime = inputs[questionNames_1.QuestionNames.Runtime] || runtime;
    }
    else {
        if ((inputs === null || inputs === void 0 ? void 0 : inputs.platform) === teamsfx_api_1.Platform.VS) {
            runtime = RuntimeOptions.DotNet().id;
        }
    }
    return runtime;
}
function botTriggerQuestion() {
    return {
        name: questionNames_1.QuestionNames.BotTrigger,
        title: localizeUtils_1.getLocalizedString("plugins.bot.questionHostTypeTrigger.title"),
        cliDescription: "Specifies the trigger for `Chat Notification Message` app template.",
        cliShortName: "t",
        type: "singleSelect",
        staticOptions: NotificationTriggerOptions.all(),
        dynamicOptions: (inputs) => {
            const runtime = getRuntime(inputs);
            return [
                runtime === RuntimeOptions.DotNet().id
                    ? NotificationTriggerOptions.appServiceForVS()
                    : NotificationTriggerOptions.appService(),
                ...NotificationTriggerOptions.functionsTriggers(),
            ];
        },
        default: (inputs) => {
            const runtime = getRuntime(inputs);
            return runtime === RuntimeOptions.DotNet().id
                ? NotificationTriggerOptions.appServiceForVS().id
                : NotificationTriggerOptions.appService().id;
        },
        placeholder: localizeUtils_1.getLocalizedString("plugins.bot.questionHostTypeTrigger.placeholder"),
    };
}
function SPFxSolutionQuestion() {
    return {
        type: "singleSelect",
        name: questionNames_1.QuestionNames.SPFxSolution,
        title: localizeUtils_1.getLocalizedString("plugins.spfx.questions.spfxSolution.title"),
        cliDescription: "Create a new or import an existing SharePoint Framework solution.",
        cliShortName: "s",
        staticOptions: [
            {
                id: "new",
                label: localizeUtils_1.getLocalizedString("plugins.spfx.questions.spfxSolution.createNew"),
                detail: localizeUtils_1.getLocalizedString("plugins.spfx.questions.spfxSolution.createNew.detail"),
            },
            {
                id: "import",
                label: localizeUtils_1.getLocalizedString("plugins.spfx.questions.spfxSolution.importExisting"),
                detail: localizeUtils_1.getLocalizedString("plugins.spfx.questions.spfxSolution.importExisting.detail"),
            },
        ],
        default: "new",
    };
}
function SPFxPackageSelectQuestion() {
    return {
        type: "singleSelect",
        name: questionNames_1.QuestionNames.SPFxInstallPackage,
        title: localizeUtils_1.getLocalizedString("plugins.spfx.questions.packageSelect.title"),
        cliDescription: "Install the latest version of SharePoint Framework.",
        staticOptions: [],
        placeholder: localizeUtils_1.getLocalizedString("plugins.spfx.questions.packageSelect.placeholder"),
        dynamicOptions: async (inputs) => {
            const versions = await Promise.all([
                utils_3.Utils.findGloballyInstalledVersion(undefined, constants_2.Constants.GeneratorPackageName, 0, false),
                utils_3.Utils.findLatestVersion(undefined, constants_2.Constants.GeneratorPackageName, 5),
                utils_3.Utils.findGloballyInstalledVersion(undefined, constants_2.Constants.YeomanPackageName, 0, false),
            ]);
            inputs.globalSpfxPackageVersion = versions[0];
            inputs.latestSpfxPackageVersion = versions[1];
            inputs.globalYeomanPackageVersion = versions[2];
            return [
                {
                    id: SPFxVersionOptionIds.installLocally,
                    label: versions[1] !== undefined
                        ? localizeUtils_1.getLocalizedString("plugins.spfx.questions.packageSelect.installLocally.withVersion.label", "v" + versions[1])
                        : localizeUtils_1.getLocalizedString("plugins.spfx.questions.packageSelect.installLocally.noVersion.label"),
                },
                {
                    id: SPFxVersionOptionIds.globalPackage,
                    label: versions[0] !== undefined
                        ? localizeUtils_1.getLocalizedString("plugins.spfx.questions.packageSelect.useGlobalPackage.withVersion.label", "v" + versions[0])
                        : localizeUtils_1.getLocalizedString("plugins.spfx.questions.packageSelect.useGlobalPackage.noVersion.label"),
                    description: localizeUtils_1.getLocalizedString("plugins.spfx.questions.packageSelect.useGlobalPackage.detail", constants_2.Constants.RecommendedLowestSpfxVersion),
                },
            ];
        },
        default: SPFxVersionOptionIds.installLocally,
        validation: {
            validFunc: (input, previousInputs) => {
                if (input === SPFxVersionOptionIds.globalPackage) {
                    const hasPackagesInstalled = !!previousInputs &&
                        !!previousInputs.globalSpfxPackageVersion &&
                        !!previousInputs.globalYeomanPackageVersion;
                    if (!hasPackagesInstalled) {
                        return Promise.reject(error_1.DevEnvironmentSetupError());
                    }
                }
                return Promise.resolve(undefined);
            },
        },
        isBoolean: true,
    };
}
exports.SPFxPackageSelectQuestion = SPFxPackageSelectQuestion;
function SPFxFrameworkQuestion() {
    return {
        type: "singleSelect",
        name: questionNames_1.QuestionNames.SPFxFramework,
        cliShortName: "k",
        cliDescription: "Framework.",
        title: localizeUtils_1.getLocalizedString("plugins.spfx.questions.framework.title"),
        staticOptions: [
            { id: "react", label: "React" },
            { id: "minimal", label: "Minimal" },
            { id: "none", label: "None" },
        ],
        placeholder: "Select an option",
        default: "react",
    };
}
exports.SPFxFrameworkQuestion = SPFxFrameworkQuestion;
function SPFxWebpartNameQuestion() {
    return {
        type: "text",
        name: questionNames_1.QuestionNames.SPFxWebpartName,
        cliShortName: "w",
        cliDescription: "Name for SharePoint Framework Web Part.",
        title: localizeUtils_1.getLocalizedString("plugins.spfx.questions.webpartName"),
        default: constants_2.Constants.DEFAULT_WEBPART_NAME,
        validation: {
            validFunc: async (input, previousInputs) => {
                const schema = {
                    pattern: "^[a-zA-Z_][a-zA-Z0-9_]*$",
                };
                const validateRes = jsonschema.validate(input, schema);
                if (validateRes.errors && validateRes.errors.length > 0) {
                    return localizeUtils_1.getLocalizedString("plugins.spfx.questions.webpartName.error.notMatch", input, schema.pattern);
                }
                if (previousInputs &&
                    ((previousInputs.stage === teamsfx_api_1.Stage.addWebpart &&
                        previousInputs[questionNames_1.QuestionNames.SPFxFolder]) ||
                        ((previousInputs === null || previousInputs === void 0 ? void 0 : previousInputs.stage) === teamsfx_api_1.Stage.addFeature && (previousInputs === null || previousInputs === void 0 ? void 0 : previousInputs.projectPath)))) {
                    const webpartFolder = path.join(previousInputs[questionNames_1.QuestionNames.SPFxFolder], "src", "webparts", input);
                    if (await fs_extra_1.default.pathExists(webpartFolder)) {
                        return localizeUtils_1.getLocalizedString("plugins.spfx.questions.webpartName.error.duplicate", webpartFolder);
                    }
                }
                return undefined;
            },
        },
    };
}
exports.SPFxWebpartNameQuestion = SPFxWebpartNameQuestion;
var SPFxVersionOptionIds;
(function (SPFxVersionOptionIds) {
    SPFxVersionOptionIds["installLocally"] = "true";
    SPFxVersionOptionIds["globalPackage"] = "false";
})(SPFxVersionOptionIds = exports.SPFxVersionOptionIds || (exports.SPFxVersionOptionIds = {}));
function SPFxImportFolderQuestion(hasDefaultFunc = false) {
    return {
        type: "folder",
        name: questionNames_1.QuestionNames.SPFxFolder,
        title: localizeUtils_1.getLocalizedString("core.spfxFolder.title"),
        cliDescription: "Directory or Path that contains the existing SharePoint Framework solution.",
        placeholder: localizeUtils_1.getLocalizedString("core.spfxFolder.placeholder"),
        default: hasDefaultFunc
            ? (inputs) => {
                if (inputs.projectPath)
                    return path.join(inputs.projectPath, "src");
                return undefined;
            }
            : undefined,
    };
}
exports.SPFxImportFolderQuestion = SPFxImportFolderQuestion;
function officeAddinHostingQuestion() {
    return {
        name: questionNames_1.QuestionNames.OfficeAddinHost,
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.officeXMLAddin.create.title"),
        type: "singleSelect",
        staticOptions: OfficeAddinHostOptions.all(),
    };
}
exports.officeAddinHostingQuestion = officeAddinHostingQuestion;
function officeAddinFrameworkQuestion() {
    return {
        type: "singleSelect",
        name: questionNames_1.QuestionNames.OfficeAddinFramework,
        cliShortName: "f",
        cliDescription: "Framework for WXP extension.",
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.officeAddin.framework.title"),
        dynamicOptions: getAddinFrameworkOptions,
        staticOptions: [
            { id: "default", label: "Default" },
            { id: "react", label: "React" },
        ],
        placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.projectType.officeAddin.framework.placeholder"),
        skipSingleOption: true,
    };
}
exports.officeAddinFrameworkQuestion = officeAddinFrameworkQuestion;
function getAddinFrameworkOptions(inputs) {
    const projectType = inputs[questionNames_1.QuestionNames.ProjectType];
    const capabilities = inputs[questionNames_1.QuestionNames.Capabilities];
    const host = inputs[questionNames_1.QuestionNames.OfficeAddinHost];
    if (projectType === ProjectTypeOptions.outlookAddin().id ||
        (projectType === ProjectTypeOptions.officeXMLAddin().id &&
            host === OfficeAddinHostOptions.outlook().id)) {
        return [{ id: "default", label: "Default" }];
    }
    else if ((projectType === ProjectTypeOptions.officeAddin().id &&
        capabilities === CapabilityOptions.officeContentAddin().id) ||
        capabilities === CapabilityOptions.officeAddinImport().id) {
        return [{ id: "default", label: "Default" }];
    }
    else {
        return [
            { id: "default", label: "Default" },
            { id: "react", label: "React" },
        ];
    }
}
exports.getAddinFrameworkOptions = getAddinFrameworkOptions;
/**
 * when project-type=office-addin-type(office-addin-framework-type=default or react), use selected value;
 * when project-type=outlook-addin-type, no framework to select, office-addin-framework-type=default_old
 * when project-type=office-xml-addin-type, no framework to select, office-addin-framework-type=default_old
 */
function getOfficeAddinFramework(inputs) {
    const projectType = inputs[questionNames_1.QuestionNames.ProjectType];
    if (projectType === ProjectTypeOptions.officeAddin().id &&
        inputs[questionNames_1.QuestionNames.OfficeAddinFramework]) {
        return inputs[questionNames_1.QuestionNames.OfficeAddinFramework];
    }
    else if ((projectType === ProjectTypeOptions.officeXMLAddin().id &&
        inputs[questionNames_1.QuestionNames.OfficeAddinHost] === OfficeAddinHostOptions.outlook().id) ||
        projectType === ProjectTypeOptions.outlookAddin().id) {
        return "default_old";
    }
    else {
        return "default";
    }
}
exports.getOfficeAddinFramework = getOfficeAddinFramework;
function getLanguageOptions(inputs) {
    const runtime = getRuntime(inputs);
    // dotnet runtime only supports C#
    if (runtime === RuntimeOptions.DotNet().id) {
        return [{ id: ProgrammingLanguage.CSharp, label: "C#" }];
    }
    const capabilities = inputs[questionNames_1.QuestionNames.Capabilities];
    const host = inputs[questionNames_1.QuestionNames.OfficeAddinHost];
    // office addin supports language defined in officeAddinJsonData
    const projectType = inputs[questionNames_1.QuestionNames.ProjectType];
    if (ProjectTypeOptions.officeAddinAllIds().includes(projectType)) {
        if (capabilities.endsWith("-manifest")) {
            return [{ id: ProgrammingLanguage.JS, label: "JavaScript" }];
        }
        if (projectType === ProjectTypeOptions.outlookAddin().id ||
            (projectType === ProjectTypeOptions.officeXMLAddin().id &&
                host === OfficeAddinHostOptions.outlook().id)) {
            return [{ id: ProgrammingLanguage.TS, label: "TypeScript" }];
        }
        const officeXMLAddinLangConfig = projectConfig_1.getOfficeAddinTemplateConfig(projectType, host)[capabilities]
            .framework["default"];
        const officeXMLAddinLangOptions = [];
        if (!!officeXMLAddinLangConfig.typescript)
            officeXMLAddinLangOptions.push({ id: ProgrammingLanguage.TS, label: "TypeScript" });
        if (!!officeXMLAddinLangConfig.javascript)
            officeXMLAddinLangOptions.push({ id: ProgrammingLanguage.JS, label: "JavaScript" });
        return officeXMLAddinLangOptions;
    }
    if (capabilities === CapabilityOptions.SPFxTab().id) {
        // SPFx only supports typescript
        return [{ id: ProgrammingLanguage.TS, label: "TypeScript" }];
    }
    else if (constants_3.capabilitiesHavePythonOption.includes(inputs[capabilities] ? inputs[capabilities] : capabilities) &&
        !(capabilities == CapabilityOptions.customCopilotRag().id &&
            (inputs[CapabilityOptions.customCopilotRag().id] ==
                CustomCopilotRagOptions.microsoft365().id ||
                inputs[CapabilityOptions.customCopilotRag().id] == CustomCopilotRagOptions.customApi().id))) {
        // support python language
        return [
            { id: ProgrammingLanguage.JS, label: "JavaScript" },
            { id: ProgrammingLanguage.TS, label: "TypeScript" },
            {
                id: ProgrammingLanguage.PY,
                label: "Python",
                detail: "",
                description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.preview"),
            },
        ];
    }
    else {
        // other cases
        return [
            { id: ProgrammingLanguage.JS, label: "JavaScript" },
            { id: ProgrammingLanguage.TS, label: "TypeScript" },
        ];
    }
}
exports.getLanguageOptions = getLanguageOptions;
var ProgrammingLanguage;
(function (ProgrammingLanguage) {
    ProgrammingLanguage["JS"] = "javascript";
    ProgrammingLanguage["TS"] = "typescript";
    ProgrammingLanguage["CSharp"] = "csharp";
    ProgrammingLanguage["PY"] = "python";
})(ProgrammingLanguage = exports.ProgrammingLanguage || (exports.ProgrammingLanguage = {}));
function programmingLanguageQuestion() {
    const programmingLanguageQuestion = {
        name: questionNames_1.QuestionNames.ProgrammingLanguage,
        cliShortName: "l",
        title: localizeUtils_1.getLocalizedString("core.ProgrammingLanguageQuestion.title"),
        type: "singleSelect",
        staticOptions: [
            { id: ProgrammingLanguage.JS, label: "JavaScript" },
            { id: ProgrammingLanguage.TS, label: "TypeScript" },
            { id: ProgrammingLanguage.CSharp, label: "C#" },
            { id: ProgrammingLanguage.PY, label: "Python" },
        ],
        dynamicOptions: getLanguageOptions,
        default: (inputs) => {
            return getLanguageOptions(inputs)[0].id;
        },
        placeholder: (inputs) => {
            const runtime = getRuntime(inputs);
            // dotnet
            if (runtime === RuntimeOptions.DotNet().id) {
                return "";
            }
            const capabilities = inputs[questionNames_1.QuestionNames.Capabilities];
            // // office addin
            // const projectType = inputs[QuestionNames.ProjectType];
            // if (projectType === ProjectTypeOptions.outlookAddin().id) {
            //   const template = getTemplate(inputs);
            //   const options = officeAddinJsonData.getSupportedScriptTypesNew(template);
            //   return options[0] || "No Options";
            // }
            // SPFx
            if (capabilities === CapabilityOptions.SPFxTab().id) {
                return localizeUtils_1.getLocalizedString("core.ProgrammingLanguageQuestion.placeholder.spfx");
            }
            // other
            return localizeUtils_1.getLocalizedString("core.ProgrammingLanguageQuestion.placeholder");
        },
        skipSingleOption: true,
    };
    return programmingLanguageQuestion;
}
exports.programmingLanguageQuestion = programmingLanguageQuestion;
function folderQuestion() {
    return {
        type: "folder",
        name: questionNames_1.QuestionNames.Folder,
        cliShortName: "f",
        title: (inputs) => teamsfx_api_1.CLIPlatforms.includes(inputs.platform)
            ? "Directory where the project folder will be created in"
            : localizeUtils_1.getLocalizedString("core.question.workspaceFolder.title"),
        cliDescription: "Directory where the project folder will be created in.",
        placeholder: localizeUtils_1.getLocalizedString("core.question.workspaceFolder.placeholder"),
        default: (inputs) => teamsfx_api_1.CLIPlatforms.includes(inputs.platform)
            ? "./"
            : path.join(os.homedir(), constants_1.ConstantString.RootFolder),
    };
}
exports.folderQuestion = folderQuestion;
exports.AppNamePattern = '^(?=(.*[\\da-zA-Z]){2})[a-zA-Z][^"<>:\\?/*&|\u0000-\u001F]*[^"\\s.<>:\\?/*&|\u0000-\u001F]$';
async function getSolutionName(spfxFolder) {
    const yoInfoPath = path.join(spfxFolder, constants_2.Constants.YO_RC_FILE);
    if (await fs_extra_1.default.pathExists(yoInfoPath)) {
        const yoInfo = await fs_extra_1.default.readJson(yoInfoPath);
        if (yoInfo["@microsoft/generator-sharepoint"]) {
            return yoInfo["@microsoft/generator-sharepoint"][constants_2.Constants.YO_RC_SOLUTION_NAME];
        }
        else {
            return undefined;
        }
    }
    else {
        throw new error_2.FileNotFoundError(constants_2.Constants.PLUGIN_NAME, yoInfoPath, constants_2.Constants.IMPORT_HELP_LINK);
    }
}
exports.getSolutionName = getSolutionName;
function appNameQuestion() {
    const question = {
        type: "text",
        name: questionNames_1.QuestionNames.AppName,
        cliShortName: "n",
        title: localizeUtils_1.getLocalizedString("core.question.appName.title"),
        required: true,
        default: async (inputs) => {
            var _a, _b;
            let defaultName = undefined;
            if ((_a = inputs.teamsAppFromTdp) === null || _a === void 0 ? void 0 : _a.appName) {
                defaultName = utils_1.convertToAlphanumericOnly((_b = inputs.teamsAppFromTdp) === null || _b === void 0 ? void 0 : _b.appName);
            }
            else if (inputs[questionNames_1.QuestionNames.SPFxSolution] == "import") {
                defaultName = await getSolutionName(inputs[questionNames_1.QuestionNames.SPFxFolder]);
            }
            else if (inputs.openAIPluginManifest) {
                defaultName = inputs.openAIPluginManifest.name_for_human;
            }
            return defaultName;
        },
        validation: {
            validFunc: async (input, previousInputs) => {
                const schema = {
                    pattern: exports.AppNamePattern,
                    maxLength: 30,
                };
                if (input.length === 25) {
                    // show warning notification because it may exceed the Teams app name max length after appending suffix
                    const context = utils_4.createContextV3();
                    if ((previousInputs === null || previousInputs === void 0 ? void 0 : previousInputs.platform) === teamsfx_api_1.Platform.VSCode) {
                        void context.userInteraction.showMessage("warn", localizeUtils_1.getLocalizedString("core.QuestionAppName.validation.lengthWarning"), false);
                    }
                    else {
                        context.logProvider.warning(localizeUtils_1.getLocalizedString("core.QuestionAppName.validation.lengthWarning"));
                    }
                }
                const appName = input;
                const validateResult = jsonschema.validate(appName, schema);
                if (validateResult.errors && validateResult.errors.length > 0) {
                    if (validateResult.errors[0].name === "pattern") {
                        return localizeUtils_1.getLocalizedString("core.QuestionAppName.validation.pattern");
                    }
                    if (validateResult.errors[0].name === "maxLength") {
                        return localizeUtils_1.getLocalizedString("core.QuestionAppName.validation.maxlength");
                    }
                }
                if (previousInputs && previousInputs.folder) {
                    const folder = previousInputs.folder;
                    if (folder) {
                        const projectPath = path.resolve(folder, appName);
                        const exists = await fs_extra_1.default.pathExists(projectPath);
                        if (exists)
                            return localizeUtils_1.getLocalizedString("core.QuestionAppName.validation.pathExist", projectPath);
                    }
                }
                return undefined;
            },
        },
        placeholder: localizeUtils_1.getLocalizedString("core.question.appName.placeholder"),
    };
    return question;
}
exports.appNameQuestion = appNameQuestion;
function sampleSelectQuestion() {
    return {
        type: "singleSelect",
        name: questionNames_1.QuestionNames.Samples,
        cliName: "sample-name",
        cliDescription: "Specifies the Microsoft Teams App sample name.",
        cliChoiceListCommand: "teamsapp list samples",
        skipValidation: true,
        cliType: "argument",
        title: localizeUtils_1.getLocalizedString("core.SampleSelect.title"),
        staticOptions: [
            "hello-world-tab-with-backend",
            "graph-toolkit-contact-exporter",
            "bot-sso",
            "todo-list-SPFx",
            "hello-world-in-meeting",
            "todo-list-with-Azure-backend-M365",
            "NPM-search-connector-M365",
            "bot-proactive-messaging-teamsfx",
            "adaptive-card-notification",
            "incoming-webhook-notification",
            "stocks-update-notification-bot",
            "query-org-user-with-message-extension-sso",
            "team-central-dashboard",
            "graph-connector-app",
            "graph-toolkit-one-productivity-hub",
            "todo-list-with-Azure-backend",
            "share-now",
            "hello-world-teams-tab-and-outlook-add-in",
            "outlook-add-in-set-signature",
            "developer-assist-dashboard",
            "live-share-dice-roller",
            "teams-chef-bot",
            "spfx-productivity-dashboard",
            "react-retail-dashboard",
            "sso-enabled-tab-via-apim-proxy",
            "large-scale-notification",
            "graph-connector-bot",
        ],
        dynamicOptions: async () => {
            return (await samples_1.sampleProvider.SampleCollection).samples.map((sample) => {
                return {
                    id: sample.id,
                    label: sample.title,
                    description: `${sample.time}  ${sample.configuration}`,
                    detail: sample.shortDescription,
                };
            });
        },
        placeholder: localizeUtils_1.getLocalizedString("core.SampleSelect.placeholder"),
        buttons: [
            {
                icon: "library",
                tooltip: localizeUtils_1.getLocalizedString("core.SampleSelect.buttons.viewSamples"),
                command: "fx-extension.openSamples",
            },
        ],
    };
}
class RuntimeOptions {
    static NodeJS() {
        return {
            id: "node",
            label: "Node.js",
            detail: localizeUtils_1.getLocalizedString("core.RuntimeOptionNodeJS.detail"),
        };
    }
    static DotNet() {
        return {
            id: "dotnet",
            label: ".NET Core",
            detail: localizeUtils_1.getLocalizedString("core.RuntimeOptionDotNet.detail"),
        };
    }
}
exports.RuntimeOptions = RuntimeOptions;
function runtimeQuestion() {
    return {
        type: "singleSelect",
        name: questionNames_1.QuestionNames.Runtime,
        title: localizeUtils_1.getLocalizedString("core.getRuntimeQuestion.title"),
        staticOptions: [RuntimeOptions.NodeJS(), RuntimeOptions.DotNet()],
        default: RuntimeOptions.NodeJS().id,
        placeholder: localizeUtils_1.getLocalizedString("core.getRuntimeQuestion.placeholder"),
        cliHidden: true,
    };
}
const defaultTabLocalHostUrl = "https://localhost:53000/index.html#/tab";
const tabContentUrlOptionItem = (tab) => {
    return {
        id: tab.name,
        label: tab.name,
        detail: localizeUtils_1.getLocalizedString("core.updateContentUrlOption.description", tab.contentUrl, defaultTabLocalHostUrl),
    };
};
const tabWebsiteUrlOptionItem = (tab) => {
    return {
        id: tab.name,
        label: tab.name,
        detail: localizeUtils_1.getLocalizedString("core.updateWebsiteUrlOption.description", tab.websiteUrl, defaultTabLocalHostUrl),
    };
};
function getTabWebsiteOptions(inputs) {
    const appDefinition = inputs.teamsAppFromTdp;
    if (appDefinition === null || appDefinition === void 0 ? void 0 : appDefinition.staticTabs) {
        const tabsWithWebsiteUrls = appDefinition.staticTabs.filter((o) => !!o.websiteUrl);
        if (tabsWithWebsiteUrls.length > 0) {
            return tabsWithWebsiteUrls.map((o) => tabWebsiteUrlOptionItem(o));
        }
    }
    return [];
}
function selectTabWebsiteUrlQuestion() {
    return {
        type: "multiSelect",
        name: questionNames_1.QuestionNames.ReplaceWebsiteUrl,
        title: localizeUtils_1.getLocalizedString("core.updateWebsiteUrlQuestion.title"),
        staticOptions: [],
        dynamicOptions: getTabWebsiteOptions,
        default: (inputs) => {
            const options = getTabWebsiteOptions(inputs);
            return options.map((o) => o.id);
        },
        placeholder: localizeUtils_1.getLocalizedString("core.updateUrlQuestion.placeholder"),
        forgetLastValue: true,
    };
}
function getTabContentUrlOptions(inputs) {
    const appDefinition = inputs.teamsAppFromTdp;
    if (appDefinition === null || appDefinition === void 0 ? void 0 : appDefinition.staticTabs) {
        const tabsWithContentUrls = appDefinition.staticTabs.filter((o) => !!o.contentUrl);
        if (tabsWithContentUrls.length > 0) {
            return tabsWithContentUrls.map((o) => tabContentUrlOptionItem(o));
        }
    }
    return [];
}
const selectTabsContentUrlQuestion = () => {
    return {
        type: "multiSelect",
        name: questionNames_1.QuestionNames.ReplaceContentUrl,
        title: localizeUtils_1.getLocalizedString("core.updateContentUrlQuestion.title"),
        staticOptions: [],
        dynamicOptions: getTabContentUrlOptions,
        default: (inputs) => {
            const options = getTabContentUrlOptions(inputs);
            return options.map((o) => o.id);
        },
        placeholder: localizeUtils_1.getLocalizedString("core.updateUrlQuestion.placeholder"),
        forgetLastValue: true,
    };
};
const answerToRepaceBotId = "bot";
const answerToReplaceMessageExtensionBotId = "messageExtension";
const botOptionItem = (isMessageExtension, botId) => {
    return {
        id: isMessageExtension ? answerToReplaceMessageExtensionBotId : answerToRepaceBotId,
        label: isMessageExtension
            ? localizeUtils_1.getLocalizedString("core.updateBotIdForMessageExtension.label")
            : localizeUtils_1.getLocalizedString("core.updateBotIdForBot.label"),
        detail: isMessageExtension
            ? localizeUtils_1.getLocalizedString("core.updateBotIdForMessageExtension.description", botId)
            : localizeUtils_1.getLocalizedString("core.updateBotIdForBot.description", botId),
    };
};
function getBotIdAndMeId(appDefinition) {
    const bots = appDefinition.bots;
    const messageExtensions = appDefinition.messagingExtensions;
    // can add only one bot. If existing, the length is 1.
    const botId = !!bots && bots.length > 0 ? bots[0].botId : undefined;
    // can add only one message extension. If existing, the length is 1.
    const messageExtensionId = !!messageExtensions && messageExtensions.length > 0 ? messageExtensions[0].botId : undefined;
    return [botId, messageExtensionId];
}
function getBotOptions(inputs) {
    const appDefinition = inputs.teamsAppFromTdp;
    if (!appDefinition)
        return [];
    const [botId, messageExtensionId] = getBotIdAndMeId(appDefinition);
    const options = [];
    if (botId) {
        options.push(botOptionItem(false, botId));
    }
    if (messageExtensionId) {
        options.push(botOptionItem(true, messageExtensionId));
    }
    return options;
}
class ApiMessageExtensionAuthOptions {
    static none() {
        return {
            id: "none",
            label: "None",
        };
    }
    static apiKey() {
        return {
            id: "api-key",
            label: "API Key",
        };
    }
    static microsoftEntra() {
        return {
            id: "microsoft-entra",
            label: "Microsoft Entra",
        };
    }
    static all() {
        return [
            ApiMessageExtensionAuthOptions.none(),
            ApiMessageExtensionAuthOptions.apiKey(),
            ApiMessageExtensionAuthOptions.microsoftEntra(),
        ];
    }
}
exports.ApiMessageExtensionAuthOptions = ApiMessageExtensionAuthOptions;
function selectBotIdsQuestion() {
    // const statcOptions: OptionItem[] = [];
    // statcOptions.push(botOptionItem(false, "000000-0000-0000"));
    // statcOptions.push(botOptionItem(true, "000000-0000-0000"));
    return {
        type: "multiSelect",
        name: questionNames_1.QuestionNames.ReplaceBotIds,
        title: localizeUtils_1.getLocalizedString("core.updateBotIdsQuestion.title"),
        staticOptions: [],
        dynamicOptions: getBotOptions,
        default: (inputs) => {
            const options = getBotOptions(inputs);
            return options.map((o) => o.id);
        },
        placeholder: localizeUtils_1.getLocalizedString("core.updateBotIdsQuestion.placeholder"),
        forgetLastValue: true,
    };
}
const maximumLengthOfDetailsErrorMessageInInputBox = 90;
function apiSpecLocationQuestion(includeExistingAPIs = true) {
    const correlationId = correlator_1.Correlator.getId(); // This is a workaround for VSCode which will lose correlation id when user accepts the value.
    const validationOnAccept = async (input, inputs) => {
        try {
            if (!inputs) {
                throw new Error("inputs is undefined"); // should never happen
            }
            const context = utils_4.createContextV3();
            const res = await helper_1.listOperations(context, undefined, input.trim(), inputs, includeExistingAPIs, false, inputs.platform === teamsfx_api_1.Platform.VSCode ? correlationId : undefined);
            if (res.isOk()) {
                inputs.supportedApisFromApiSpec = res.value;
            }
            else {
                const errors = res.error;
                if (inputs.platform === teamsfx_api_1.Platform.CLI) {
                    return errors.map((e) => e.content).join("\n");
                }
                if (errors.length === 1 &&
                    errors[0].content.length <= maximumLengthOfDetailsErrorMessageInInputBox) {
                    return errors[0].content;
                }
                else {
                    return localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpec.multipleValidationErrors.vscode.message");
                }
            }
        }
        catch (e) {
            const error = error_2.assembleError(e);
            throw error;
        }
    };
    return {
        type: "singleFileOrText",
        name: questionNames_1.QuestionNames.ApiSpecLocation,
        cliShortName: "a",
        cliDescription: "OpenAPI description document location.",
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpec.title"),
        forgetLastValue: true,
        inputBoxConfig: {
            type: "innerText",
            title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpec.title"),
            placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpec.placeholder"),
            name: "input-api-spec-url",
            step: 2,
            validation: {
                validFunc: (input, inputs) => {
                    const result = util_1.isValidHttpUrl(input.trim())
                        ? undefined
                        : (inputs === null || inputs === void 0 ? void 0 : inputs.platform) === teamsfx_api_1.Platform.CLI
                            ? "Please enter a valid HTTP URL without authentication to access your OpenAPI description document or enter a file path of your local OpenAPI description document."
                            : localizeUtils_1.getLocalizedString("core.createProjectQuestion.invalidUrl.message");
                    return Promise.resolve(result);
                },
            },
        },
        inputOptionItem: {
            id: "input",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpecInputUrl.label"),
        },
        filters: {
            files: ["json", "yml", "yaml"],
        },
        validation: {
            validFunc: async (input, inputs) => {
                if (!util_1.isValidHttpUrl(input.trim()) && !(await fs_extra_1.default.pathExists(input.trim()))) {
                    return "Please enter a valid HTTP URL without authentication to access your OpenAPI description document or enter a file path of your local OpenAPI description document.";
                }
                return await validationOnAccept(input, inputs);
            },
        },
    };
}
exports.apiSpecLocationQuestion = apiSpecLocationQuestion;
function openAIPluginManifestLocationQuestion() {
    // export for unit test
    const correlationId = correlator_1.Correlator.getId(); // This is a workaround for VSCode which will lose correlation id when user accepts the value.
    return {
        type: "text",
        name: questionNames_1.QuestionNames.OpenAIPluginManifest,
        cliShortName: "m",
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.OpenAIPluginDomain"),
        placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.OpenAIPluginDomain.placeholder"),
        cliDescription: "OpenAI plugin website domain or manifest URL.",
        forgetLastValue: true,
        validation: {
            validFunc: (input) => {
                const pattern = /(https?:\/\/)?([a-z0-9-]+(\.[a-z0-9-]+)*)(:[0-9]{1,5})?(\/)?$/i;
                const match = pattern.test(input);
                const result = match
                    ? undefined
                    : localizeUtils_1.getLocalizedString("core.createProjectQuestion.invalidUrl.message");
                return Promise.resolve(result);
            },
        },
        additionalValidationOnAccept: {
            validFunc: async (input, inputs) => {
                if (!inputs) {
                    throw new Error("inputs is undefined"); // should never happen
                }
                let manifest;
                try {
                    manifest = await helper_1.OpenAIPluginManifestHelper.loadOpenAIPluginManifest(input);
                    inputs.openAIPluginManifest = manifest;
                }
                catch (e) {
                    const error = error_2.assembleError(e);
                    return error.message;
                }
                const context = utils_4.createContextV3();
                try {
                    const res = await helper_1.listOperations(context, manifest, inputs[questionNames_1.QuestionNames.ApiSpecLocation], inputs, true, true, inputs.platform === teamsfx_api_1.Platform.VSCode ? correlationId : undefined);
                    if (res.isOk()) {
                        inputs.supportedApisFromApiSpec = res.value;
                    }
                    else {
                        const errors = res.error;
                        if (inputs.platform === teamsfx_api_1.Platform.CLI) {
                            return errors.map((e) => e.content).join("\n");
                        }
                        if (errors.length === 1 &&
                            errors[0].content.length <= maximumLengthOfDetailsErrorMessageInInputBox) {
                            return errors[0].content;
                        }
                        else {
                            return localizeUtils_1.getLocalizedString("core.createProjectQuestion.openAiPluginManifest.multipleValidationErrors.vscode.message");
                        }
                    }
                }
                catch (e) {
                    const error = error_2.assembleError(e);
                    throw error;
                }
            },
        },
    };
}
exports.openAIPluginManifestLocationQuestion = openAIPluginManifestLocationQuestion;
function apiMessageExtensionAuthQuestion() {
    return {
        type: "singleSelect",
        name: questionNames_1.QuestionNames.ApiMEAuth,
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiMessageExtensionAuth.title"),
        placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiMessageExtensionAuth.placeholder"),
        cliDescription: "The authentication type for the API.",
        staticOptions: ApiMessageExtensionAuthOptions.all(),
        dynamicOptions: () => ApiMessageExtensionAuthOptions.all(),
        default: ApiMessageExtensionAuthOptions.none().id,
    };
}
exports.apiMessageExtensionAuthQuestion = apiMessageExtensionAuthQuestion;
function apiOperationQuestion(includeExistingAPIs = true, isAddPlugin = false) {
    // export for unit test
    let placeholder = "";
    const isPlugin = (inputs) => {
        return (isAddPlugin ||
            (!!inputs &&
                inputs[questionNames_1.QuestionNames.Capabilities] === CapabilityOptions.copilotPluginApiSpec().id));
    };
    return {
        type: "multiSelect",
        name: questionNames_1.QuestionNames.ApiOperation,
        title: (inputs) => {
            return isPlugin(inputs)
                ? localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpec.copilotOperation.title")
                : localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpec.operation.title");
        },
        cliDescription: isAddPlugin
            ? "Select operation(s) Copilot can interact with."
            : "Select operation(s) Teams can interact with.",
        cliShortName: "o",
        placeholder: (inputs) => {
            const isPlugin = inputs[questionNames_1.QuestionNames.Capabilities] === CapabilityOptions.copilotPluginApiSpec().id;
            if (!includeExistingAPIs) {
                placeholder = localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpec.operation.placeholder.skipExisting");
            }
            else if (isPlugin) {
                placeholder = ""; // TODO: add placeholder for api plugin
            }
            else {
                placeholder = localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpec.operation.apikey.placeholder");
            }
            return placeholder;
        },
        forgetLastValue: true,
        staticOptions: [],
        validation: {
            validFunc: (input, inputs) => {
                if (!inputs) {
                    throw new Error("inputs is undefined"); // should never happen
                }
                if (input.length < 1 ||
                    (input.length > 10 &&
                        inputs[questionNames_1.QuestionNames.CustomCopilotRag] != CustomCopilotRagOptions.customApi().id)) {
                    return localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpec.operation.invalidMessage", input.length, 10);
                }
                const operations = inputs.supportedApisFromApiSpec;
                const authNames = new Set();
                const serverUrls = new Set();
                for (const inputItem of input) {
                    const operation = operations.find((op) => op.id === inputItem);
                    if (operation) {
                        if (operation.data.authName) {
                            authNames.add(operation.data.authName);
                            serverUrls.add(operation.data.serverUrl);
                        }
                    }
                }
                if (authNames.size > 1) {
                    return localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpec.operation.multipleAuth", Array.from(authNames).join(", "));
                }
                if (serverUrls.size > 1) {
                    return localizeUtils_1.getLocalizedString("core.createProjectQuestion.apiSpec.operation.multipleServer", Array.from(serverUrls).join(", "));
                }
                const authApi = operations.find((api) => !!api.data.authName && input.includes(api.id));
                if (authApi) {
                    inputs.apiAuthData = authApi.data;
                }
            },
        },
        dynamicOptions: (inputs) => {
            if (!inputs.supportedApisFromApiSpec) {
                throw new error_2.EmptyOptionError(questionNames_1.QuestionNames.ApiOperation, "question");
            }
            const operations = inputs.supportedApisFromApiSpec;
            return operations;
        },
    };
}
exports.apiOperationQuestion = apiOperationQuestion;
class CustomCopilotRagOptions {
    static customize() {
        return {
            id: "custom-copilot-rag-customize",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRagCustomizeOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRagCustomizeOption.detail"),
        };
    }
    static azureAISearch() {
        return {
            id: "custom-copilot-rag-azureAISearch",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRagAzureAISearchOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRagAzureAISearchOption.detail"),
        };
    }
    static customApi() {
        return {
            id: "custom-copilot-rag-customApi",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRagCustomApiOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRagCustomApiOption.detail"),
            description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.preview"),
        };
    }
    static microsoft365() {
        return {
            id: "custom-copilot-rag-microsoft365",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRagMicrosoft365Option.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRagMicrosoft365Option.detail"),
        };
    }
    static all() {
        return [
            CustomCopilotRagOptions.customize(),
            CustomCopilotRagOptions.azureAISearch(),
            CustomCopilotRagOptions.customApi(),
            CustomCopilotRagOptions.microsoft365(),
        ];
    }
}
exports.CustomCopilotRagOptions = CustomCopilotRagOptions;
class CustomCopilotAssistantOptions {
    static new() {
        return {
            id: "custom-copilot-agent-new",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotAssistantNewOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotAssistantNewOption.detail"),
        };
    }
    static assistantsApi() {
        return {
            id: "custom-copilot-agent-assistants-api",
            label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotAssistantAssistantsApiOption.label"),
            detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotAssistantAssistantsApiOption.detail"),
            description: localizeUtils_1.getLocalizedString("core.createProjectQuestion.option.description.preview"),
        };
    }
    static all() {
        return [CustomCopilotAssistantOptions.new(), CustomCopilotAssistantOptions.assistantsApi()];
    }
}
exports.CustomCopilotAssistantOptions = CustomCopilotAssistantOptions;
function customCopilotRagQuestion() {
    return {
        type: "singleSelect",
        name: questionNames_1.QuestionNames.CustomCopilotRag,
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRag.title"),
        placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotRag.placeholder"),
        staticOptions: CustomCopilotRagOptions.all(),
        dynamicOptions: () => CustomCopilotRagOptions.all(),
        default: CustomCopilotRagOptions.customize().id,
    };
}
function customCopilotAssistantQuestion() {
    return {
        type: "singleSelect",
        name: questionNames_1.QuestionNames.CustomCopilotAssistant,
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotAssistant.title"),
        placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.capability.customCopilotAssistant.placeholder"),
        staticOptions: CustomCopilotAssistantOptions.all(),
        dynamicOptions: () => CustomCopilotAssistantOptions.all(),
        default: CustomCopilotAssistantOptions.new().id,
    };
}
function llmServiceQuestion() {
    return {
        type: "singleSelect",
        name: questionNames_1.QuestionNames.LLMService,
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmService.title"),
        placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmService.placeholder"),
        staticOptions: [
            {
                id: "llm-service-azure-openai",
                cliName: "azure-openai",
                label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmServiceAzureOpenAIOption.label"),
                detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmServiceAzureOpenAIOption.detail"),
            },
            {
                id: "llm-service-openai",
                label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmServiceOpenAIOption.label"),
                detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmServiceOpenAIOption.detail"),
            },
        ],
        dynamicOptions: (inputs) => {
            const options = [];
            if (inputs[questionNames_1.QuestionNames.CustomCopilotAssistant] !== "custom-copilot-agent-assistants-api") {
                options.push({
                    id: "llm-service-azure-openai",
                    label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmServiceAzureOpenAIOption.label"),
                    detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmServiceAzureOpenAIOption.detail"),
                });
            }
            options.push({
                id: "llm-service-openai",
                label: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmServiceOpenAIOption.label"),
                detail: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmServiceOpenAIOption.detail"),
            });
            return options;
        },
        skipSingleOption: true,
        default: "llm-service-azure-openai",
    };
}
function openAIKeyQuestion() {
    return {
        type: "text",
        password: true,
        name: questionNames_1.QuestionNames.OpenAIKey,
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmService.openAIKey.title"),
        placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmService.openAIKey.placeholder"),
    };
}
function azureOpenAIKeyQuestion() {
    return {
        type: "text",
        password: true,
        name: questionNames_1.QuestionNames.AzureOpenAIKey,
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmService.azureOpenAIKey.title"),
        placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmService.azureOpenAIKey.placeholder"),
    };
}
function azureOpenAIEndpointQuestion() {
    return {
        type: "text",
        name: questionNames_1.QuestionNames.AzureOpenAIEndpoint,
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmService.azureOpenAIEndpoint.title"),
        placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmService.azureOpenAIEndpoint.placeholder"),
    };
}
function azureOpenAIDeploymentNameQuestion() {
    return {
        type: "text",
        name: questionNames_1.QuestionNames.AzureOpenAIDeploymentName,
        title: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmService.azureOpenAIDeploymentName.title"),
        placeholder: localizeUtils_1.getLocalizedString("core.createProjectQuestion.llmService.azureOpenAIDeploymentName.placeholder"),
    };
}
function capabilitySubTree() {
    const node = {
        data: capabilityQuestion(),
        children: [
            {
                // Notification bot trigger sub-tree
                condition: { equals: CapabilityOptions.notificationBot().id },
                data: botTriggerQuestion(),
            },
            {
                // SPFx sub-tree
                condition: { equals: CapabilityOptions.SPFxTab().id },
                data: SPFxSolutionQuestion(),
                children: [
                    {
                        data: { type: "group" },
                        children: [
                            { data: SPFxPackageSelectQuestion() },
                            { data: SPFxFrameworkQuestion() },
                            { data: SPFxWebpartNameQuestion() },
                        ],
                        condition: { equals: "new" },
                    },
                    {
                        data: SPFxImportFolderQuestion(),
                        condition: { equals: "import" },
                    },
                ],
            },
            {
                // office addin import sub-tree (capabilities=office-addin-import | outlook-addin-import)
                condition: {
                    enum: [
                        CapabilityOptions.outlookAddinImport().id,
                        CapabilityOptions.officeAddinImport().id,
                    ],
                },
                data: { type: "group", name: questionNames_1.QuestionNames.OfficeAddinImport },
                children: [
                    {
                        data: {
                            type: "folder",
                            name: questionNames_1.QuestionNames.OfficeAddinFolder,
                            title: "Existing add-in project folder",
                        },
                    },
                    {
                        data: {
                            type: "singleFile",
                            name: questionNames_1.QuestionNames.OfficeAddinManifest,
                            title: "Select import project manifest file",
                        },
                    },
                ],
            },
            {
                // Search ME sub-tree
                condition: { equals: CapabilityOptions.m365SearchMe().id },
                data: meArchitectureQuestion(),
            },
            {
                // API ME from API Spec or Copilot plugin from API spec or AI Plugin
                condition: (inputs) => {
                    return (inputs[questionNames_1.QuestionNames.Capabilities] === CapabilityOptions.copilotPluginApiSpec().id ||
                        inputs[questionNames_1.QuestionNames.Capabilities] ===
                            CapabilityOptions.copilotPluginOpenAIPlugin().id ||
                        inputs[questionNames_1.QuestionNames.MeArchitectureType] === MeArchitectureOptions.apiSpec().id);
                },
                data: { type: "group", name: questionNames_1.QuestionNames.CopilotPluginExistingApi },
                children: [
                    {
                        data: apiSpecLocationQuestion(),
                    },
                    // {
                    //   condition: { equals: CapabilityOptions.copilotPluginOpenAIPlugin().id },
                    //   data: openAIPluginManifestLocationQuestion(),
                    // },
                    {
                        data: apiOperationQuestion(),
                    },
                ],
            },
            {
                condition: (inputs) => {
                    return inputs[questionNames_1.QuestionNames.MeArchitectureType] == MeArchitectureOptions.newApi().id;
                },
                data: apiMessageExtensionAuthQuestion(),
            },
            {
                condition: (inputs) => {
                    return inputs[questionNames_1.QuestionNames.Capabilities] == CapabilityOptions.customCopilotRag().id;
                },
                data: customCopilotRagQuestion(),
                children: [
                    {
                        condition: (inputs) => {
                            return (inputs[questionNames_1.QuestionNames.CustomCopilotRag] === CustomCopilotRagOptions.customApi().id);
                        },
                        data: apiSpecLocationQuestion(),
                    },
                    {
                        condition: (inputs) => {
                            return (inputs[questionNames_1.QuestionNames.CustomCopilotRag] === CustomCopilotRagOptions.customApi().id);
                        },
                        data: apiOperationQuestion(),
                    },
                ],
            },
            {
                condition: (inputs) => {
                    return (inputs[questionNames_1.QuestionNames.Capabilities] == CapabilityOptions.customCopilotAssistant().id);
                },
                data: customCopilotAssistantQuestion(),
            },
            {
                // programming language
                data: programmingLanguageQuestion(),
                condition: (inputs) => {
                    return (!!inputs[questionNames_1.QuestionNames.Capabilities] &&
                        inputs[questionNames_1.QuestionNames.Capabilities] !== CapabilityOptions.copilotPluginApiSpec().id &&
                        inputs[questionNames_1.QuestionNames.Capabilities] !==
                            CapabilityOptions.copilotPluginOpenAIPlugin().id &&
                        inputs[questionNames_1.QuestionNames.Capabilities] !== CapabilityOptions.customizeGptBasic().id &&
                        inputs[questionNames_1.QuestionNames.MeArchitectureType] !== MeArchitectureOptions.apiSpec().id &&
                        inputs[questionNames_1.QuestionNames.Capabilities] !== CapabilityOptions.officeAddinImport().id &&
                        inputs[questionNames_1.QuestionNames.Capabilities] !== CapabilityOptions.outlookAddinImport().id);
                },
            },
            {
                condition: (inputs) => {
                    return (inputs[questionNames_1.QuestionNames.Capabilities] === CapabilityOptions.customCopilotBasic().id ||
                        inputs[questionNames_1.QuestionNames.Capabilities] === CapabilityOptions.customCopilotRag().id ||
                        inputs[questionNames_1.QuestionNames.Capabilities] === CapabilityOptions.customCopilotAssistant().id);
                },
                data: llmServiceQuestion(),
                children: [
                    {
                        condition: { equals: "llm-service-azure-openai" },
                        data: azureOpenAIKeyQuestion(),
                        children: [
                            {
                                condition: (inputs) => {
                                    var _a;
                                    return ((_a = inputs[questionNames_1.QuestionNames.AzureOpenAIKey]) === null || _a === void 0 ? void 0 : _a.length) > 0;
                                },
                                data: azureOpenAIEndpointQuestion(),
                                children: [
                                    {
                                        condition: (inputs) => {
                                            var _a;
                                            return ((_a = inputs[questionNames_1.QuestionNames.AzureOpenAIEndpoint]) === null || _a === void 0 ? void 0 : _a.length) > 0;
                                        },
                                        data: azureOpenAIDeploymentNameQuestion(),
                                    },
                                ],
                            },
                        ],
                    },
                    {
                        condition: { equals: "llm-service-openai" },
                        data: openAIKeyQuestion(),
                    },
                ],
            },
            {
                // Office addin framework for json manifest
                data: officeAddinFrameworkQuestion(),
                condition: (inputs) => {
                    return (inputs[questionNames_1.QuestionNames.ProjectType] === ProjectTypeOptions.officeAddin().id &&
                        inputs[questionNames_1.QuestionNames.Capabilities] !== CapabilityOptions.officeAddinImport().id);
                },
            },
            {
                // root folder
                data: folderQuestion(),
            },
            {
                // app name
                data: appNameQuestion(),
            },
        ],
        condition: (inputs) => {
            return inputs[questionNames_1.QuestionNames.ProjectType] !== ProjectTypeOptions.startWithGithubCopilot().id;
        },
    };
    return node;
}
exports.capabilitySubTree = capabilitySubTree;
function createProjectQuestionNode() {
    const createProjectQuestion = {
        data: { type: "group" },
        children: [
            {
                condition: (inputs) => featureFlags_1.isCLIDotNetEnabled() && teamsfx_api_1.CLIPlatforms.includes(inputs.platform),
                data: runtimeQuestion(),
            },
            {
                condition: (inputs) => inputs.platform === teamsfx_api_1.Platform.VSCode || inputs.platform === teamsfx_api_1.Platform.CLI,
                data: projectTypeQuestion(),
                cliOptionDisabled: "self",
            },
            {
                condition: (inputs) => inputs[questionNames_1.QuestionNames.ProjectType] === ProjectTypeOptions.officeXMLAddin().id,
                data: officeAddinHostingQuestion(),
            },
            capabilitySubTree(),
            {
                condition: (inputs) => inputs.teamsAppFromTdp && utils_2.isPersonalApp(inputs.teamsAppFromTdp),
                data: { type: "group", name: questionNames_1.QuestionNames.RepalceTabUrl },
                cliOptionDisabled: "all",
                inputsDisabled: "all",
                children: [
                    {
                        condition: (inputs) => {
                            var _a;
                            return (((_a = inputs.teamsAppFromTdp) === null || _a === void 0 ? void 0 : _a.staticTabs.filter((o) => !!o.websiteUrl)) || []).length >
                                0;
                        },
                        data: selectTabWebsiteUrlQuestion(),
                    },
                    {
                        condition: (inputs) => {
                            var _a;
                            return (((_a = inputs.teamsAppFromTdp) === null || _a === void 0 ? void 0 : _a.staticTabs.filter((o) => !!o.contentUrl)) || []).length >
                                0;
                        },
                        data: selectTabsContentUrlQuestion(),
                    },
                ],
            },
            {
                condition: (inputs) => {
                    const appDef = inputs.teamsAppFromTdp;
                    return appDef && utils_2.needBotCode(appDef);
                },
                data: selectBotIdsQuestion(),
                cliOptionDisabled: "all",
                inputsDisabled: "all",
            },
        ],
    };
    return createProjectQuestion;
}
exports.createProjectQuestionNode = createProjectQuestionNode;
function createSampleProjectQuestionNode() {
    return {
        data: sampleSelectQuestion(),
        children: [
            {
                data: folderQuestion(),
            },
        ],
    };
}
exports.createSampleProjectQuestionNode = createSampleProjectQuestionNode;
function createProjectCliHelpNode() {
    const node = lodash_1.cloneDeep(createProjectQuestionNode());
    const deleteNames = [
        questionNames_1.QuestionNames.ProjectType,
        questionNames_1.QuestionNames.OfficeAddinImport,
        questionNames_1.QuestionNames.OfficeAddinHost,
        questionNames_1.QuestionNames.RepalceTabUrl,
        questionNames_1.QuestionNames.ReplaceBotIds,
        questionNames_1.QuestionNames.Samples,
    ];
    if (!featureFlags_1.isCLIDotNetEnabled()) {
        deleteNames.push(questionNames_1.QuestionNames.Runtime);
    }
    trimQuestionTreeForCliHelp(node, deleteNames);
    return node;
}
exports.createProjectCliHelpNode = createProjectCliHelpNode;
function trimQuestionTreeForCliHelp(node, deleteNames) {
    if (node.children) {
        node.children = node.children.filter((child) => !child.data.name || !deleteNames.includes(child.data.name));
        for (const child of node.children) {
            trimQuestionTreeForCliHelp(child, deleteNames);
        }
    }
}
function pickSubTree(node, name) {
    if (node.data.name === name) {
        return node;
    }
    let found;
    if (node.children) {
        for (const child of node.children) {
            found = pickSubTree(child, name);
            if (found)
                return found;
        }
    }
    return undefined;
}
