// @microsoft/teamsfx-core/build/component/configManager/actionInjector.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionInjector = void 0;
const tslib_1 = require("tslib");
const m365_spec_parser_1 = require("@microsoft/m365-spec-parser");
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const yaml_1 = require("yaml");
const common_1 = require("../../error/common");
class ActionInjector {
    static hasActionWithName(provisionNode, action, name) {
        const hasAuthAction = provisionNode.items.some((item) => { var _a; return item.get("uses") === action && ((_a = item.get("with")) === null || _a === void 0 ? void 0 : _a.get("name")) === name; });
        return hasAuthAction;
    }
    static getTeamsAppIdEnvName(provisionNode) {
        var _a;
        for (const item of provisionNode.items) {
            if (item.get("uses") === "teamsApp/create") {
                return (_a = item.get("writeToEnvironmentFile")) === null || _a === void 0 ? void 0 : _a.get("teamsAppId");
            }
        }
        return undefined;
    }
    static generateAuthAction(actionName, authName, teamsAppIdEnvName, specRelativePath, envName, flow) {
        const result = {
            uses: actionName,
            with: {
                name: `${authName}`,
                appId: `\${{${teamsAppIdEnvName}}}`,
                apiSpecPath: specRelativePath,
            },
        };
        if (flow) {
            result.with.flow = flow;
            result.writeToEnvironmentFile = {
                configurationId: envName,
            };
        }
        else {
            result.writeToEnvironmentFile = {
                registrationId: envName,
            };
        }
        return result;
    }
    static async injectCreateOAuthAction(ymlPath, authName, specRelativePath) {
        const ymlContent = await fs_extra_1.default.readFile(ymlPath, "utf-8");
        const actionName = "oauth/register";
        const document = yaml_1.parseDocument(ymlContent);
        const provisionNode = document.get("provision");
        if (provisionNode) {
            const hasOAuthAction = ActionInjector.hasActionWithName(provisionNode, actionName, authName);
            if (!hasOAuthAction) {
                provisionNode.items = provisionNode.items.filter((item) => item.get("uses") !== actionName && item.get("uses") !== "apiKey/register");
                const teamsAppIdEnvName = ActionInjector.getTeamsAppIdEnvName(provisionNode);
                if (teamsAppIdEnvName) {
                    const index = provisionNode.items.findIndex((item) => item.get("uses") === "teamsApp/create");
                    const envName = m365_spec_parser_1.Utils.getSafeRegistrationIdEnvName(`${authName}_CONFIGURATION_ID`);
                    const flow = "authorizationCode";
                    const action = ActionInjector.generateAuthAction(actionName, authName, teamsAppIdEnvName, specRelativePath, envName, flow);
                    provisionNode.items.splice(index + 1, 0, action);
                }
                else {
                    throw new common_1.InjectOAuthActionFailedError();
                }
                await fs_extra_1.default.writeFile(ymlPath, document.toString(), "utf8");
            }
        }
        else {
            throw new common_1.InjectOAuthActionFailedError();
        }
    }
    static async injectCreateAPIKeyAction(ymlPath, authName, specRelativePath) {
        const ymlContent = await fs_extra_1.default.readFile(ymlPath, "utf-8");
        const actionName = "apiKey/register";
        const document = yaml_1.parseDocument(ymlContent);
        const provisionNode = document.get("provision");
        if (provisionNode) {
            const hasApiKeyAction = ActionInjector.hasActionWithName(provisionNode, actionName, authName);
            if (!hasApiKeyAction) {
                provisionNode.items = provisionNode.items.filter((item) => item.get("uses") !== actionName && item.get("uses") !== "oauth/register");
                const teamsAppIdEnvName = ActionInjector.getTeamsAppIdEnvName(provisionNode);
                if (teamsAppIdEnvName) {
                    const index = provisionNode.items.findIndex((item) => item.get("uses") === "teamsApp/create");
                    const envName = m365_spec_parser_1.Utils.getSafeRegistrationIdEnvName(`${authName}_REGISTRATION_ID`);
                    const action = ActionInjector.generateAuthAction(actionName, authName, teamsAppIdEnvName, specRelativePath, envName);
                    provisionNode.items.splice(index + 1, 0, action);
                }
                else {
                    throw new common_1.InjectAPIKeyActionFailedError();
                }
                await fs_extra_1.default.writeFile(ymlPath, document.toString(), "utf8");
            }
        }
        else {
            throw new common_1.InjectAPIKeyActionFailedError();
        }
    }
}
exports.ActionInjector = ActionInjector;
